{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e525f275",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO \n",
    "import os,sys\n",
    "import warnings   \n",
    "warnings.filterwarnings('ignore')\n",
    "import configparser\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import argparse\n",
    "import json\n",
    "from os.path import exists,splitext,dirname,splitext,basename,realpath,abspath\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d588ca7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gb_file = './input/'+ 'GCF_000005845.2_ASM584v2_genomic.gbff'\n",
    "input = pd.read_excel('./input/基因敲除目标.xlsx',sheet_name='ID')\n",
    "id = input['GeneID']\n",
    "gb1_file = '/home/yanghe/program/data_preprocessing/input/参考基因组eco.gb'\n",
    "gb2_file = '/home/yanghe/program/data_preprocessing/input/GCF_000146045.2_R64_genomic.gbff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d5c51f64",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/home/yanghe/program/data_preprocessing/input/05-10-input.csv')\n",
    "df = df.drop(columns=['Unnamed: 6','Unnamed: 7','Unnamed: 8','Unnamed: 9','Unnamed: 10','Unnamed: 11','Unnamed: 12'])\n",
    "df = df.append(df).append(df).append(df).append(df)\n",
    "deletion = df.copy()\n",
    "\n",
    "df['Up region'] = 20\n",
    "df['Inserted sequence'] = 'agctcgatcctcat'\n",
    "df['Manipulation type'] = 'insertion'\n",
    "insertion = df.copy()\n",
    "\n",
    "df['Up region'] = 20\n",
    "df['Down region'] = 100\n",
    "df['Inserted sequence'] = 'agctcgatcctcatagctcgatcctcatagctcgatcctcatagctcgatcctcatagctcgatcctcat'\n",
    "df['Manipulation type'] = 'substitution'\n",
    "substitution = df.copy()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "86424550",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('05-14-input.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0e8718b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = deletion.append(insertion).append(substitution)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5472553f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2f0e119f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def work(record, gene_id):\n",
    "    # 读入GenBank格式的文件\n",
    "   \n",
    "    # 提取基因序列和其位置\n",
    "    gene_id = gene_id  # 将YOUR_GENE_ID替换为你要提取的基因ID\n",
    "    \n",
    "    for feature in record.features:\n",
    "        if feature.type == \"gene\" and gene_id in feature.qualifiers.get(\"locus_tag\", []):\n",
    "            #取序列\n",
    "            gene_seq = feature.extract(record.seq)\n",
    "            if feature.strand == -1:  # 如果是反义链，则需要取反向互补序列\n",
    "                gene_seq = gene_seq.reverse_complement()\n",
    "            \n",
    "            #取坐标\n",
    "            gene_start = feature.location.start.position\n",
    "            gene_end = feature.location.end.position\n",
    "            \n",
    "            #取目标序列的前100bp\n",
    "            start_pos = max(gene_start - 100, 0)\n",
    "            before_gene_seq = record.seq[start_pos:gene_start]\n",
    "            \n",
    "            print(f\">{gene_id} {record.id}:{gene_start}-{gene_end}\\n{gene_seq}\")\n",
    "            break\n",
    "    return gene_id, gene_seq, before_gene_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1ecf02d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">b3956 NC_000913.3:4150446-4153098\n",
      "TTAGCCGGTATTACGCATACCTGCCGCAATCCCGGCAATAGTGACCATTAACGCTTGTTCGACGCGAGGATCCGGTTCCTGGCCTTCTTTTTCTGCCTGGCGGGAGCGGTGCAGCAACTCGGCCTGCAATACGTTCAGCGGGTCGGTGTAAATATTCCGTAGCTGAATAGACTCTGCAATCCACGGCAGATCGGCCATCAGATGGGAATCGTTGGCAATCGCCAGCACCACTTTGATGTCTTCTTCTTGCAGGTTGCGTAACTCTTTACCTAACGGCCACAGTGCTTTGTCTACCAGGCGTTGGTCATAGTATTCCGCCAGCCACAGGTCTGCTTTGGCGAAGACCATCTCCAGCATGCCGAGACGCGTCGAGAAGAATGGCCAATCGCGGCACATAGCCTCCAGCTCGCTCTGTTTGCCGTCTTCGACCACTTTTTGCAGCGCCGTACCTGCACCCAGCCAGGCGGGGAGCATCAGACGGTTTTGCGTCCAGGCGAAGATCCACGGAATGGCGCGTAGTGACTCGACGCCGCCGGTTGGGCGACGTTTCGCCGGACGTGAACCCAACGGCAGTTTGCCCAGTTCTTGTTCCGGCGTAGCGGAGCGGAAGTAAGGCACAAAATCTTTGTTTTCACGTACGTAGCCGCGGTAGACATCGCAGGAGATGACTGACAGTTCATCCATAATGCGACGCCAGCTCTCTTTCGGCTCCGGCGGTGGCAGCAGGTTGGCTTCCAGAATCGCCCCGGTATAAAGCGACAGGCTGCTGACGGTGATTTCTGGCAGACCATATTTAAAGCGGATCATCTCGCCCTGTTCGGTTACGCGCAGGCCGCCTTTCAGGCTTCCTGGCGGTTGTGACAGCAGCGCCGCATGAGCAGGTGCGCCGCCGCGACCAATGGAACCGCCGCGACCGTGGAACAACGTCAGCTCAATACCCGCTTTTTCGCAGGTTTTGATTAATGCATCCTGTGCCTGATATTGCGCCCAGGAAGCTGCCATCACTCCCGCATCTTTTGCTGAGTCGGAATAGCCAATCATCACCATCTGTTTGCCCTGAATCAGGCCACGATACCAGTCAATATTGAGCAGCTGGGTCATGACATCGTTGGCGTTGTTCAGATCATCGAGGGTTTCAAACAGCGGAGCAACCGGCATCGCAAACCCGATACCCGCTTCTTTCAGCAGCAGGTGGACAGCCAGTACGTCGGACGGCGTTTTCGCCATCGAGATCACGTAGGCGGCAATGGAGCCTTGCGGTGCTTCGGCAATCACCTGGCAGGTATCGAGCACTTCGCGCGTTTCGGCGCTTGGTTGCCAGTTGCGCGGCAGAAGCGGACGTTTGGAGTTCAGTTCGCGGATCAGGAACGCCTGTTTGTCGGCCTCTGACCAGCTTTCGTAGTCGCCGATACCGAGGTAGCGGGTCAGCTCGCCCAGCGCTTCGGTATGACGCGTGCTCTCCTGACGGATATCAATACGGACCAGCGGTACGCCGAAACATTTCACGCGGCGCAGGGTGTCGAGCAGATCGCCGTTGGCGATAATACCCATGCCACACGCCTGAAGTGACTGGTAGCAAGCGTAGAGCGGTTCCCACAGTTCTTCGTTTTGTGTCAGCAGGCCTTCTGGTTTTGGCAGTTCTTCGCCTTTCAGGCGCGCTTCCAGCCATGCCTGTGTCGCCATCAGGCGAGAACGCAGGTTTTTCATCAGATAGCGATACGGTTCTGCGGCACCTTCTTCGCCAACCAGCGCCAGCAGTTCAGGGGTCGCTTCAACCATCGACAGTTCAGAAACCAGCACCTGAATATCTTTCAGGAACAAATCGGTGGCTTTCCAGCGGCTGAGTAGCAGGACGTGGCGGGTGATATCGGCAGTGACGTTCGGGTTGCCGTCGCGGTCGCCGCCCATCCACGAAGTAAAACGGACCGGAACAAATTCGACGGGCAGTTTGTAGCCGAGGTTCTCTTCCAGTTGTTCGTTCAGTTCGCGCAGGTAATTTGGTACGCCTTGCCACAGGCTGTTTTCCACTACGGCAAAGCCCCATTTGGCTTCATCTACCGGGCTTGGACGCAGCTTACGGATTTCATCGGTATGCCATGACTGGGCGATCAACTGGCGCAGGCGACGCATCAGCTGGTTGTGTTCGTAGTCAGCGATATCTTTGTTATCGAGCTGTTTTAAACAGGCGTTCACTTCCACCATTTTGTGGATCAGTGTACGACGGGTAATTTCGGTTGGGTGAGCCGTGAGGACCAGTTCCAGCGACAGCGATTCCACTGCTTTTTTGATGGTGTCTTCGCTCAGTTCCGGCTGGTTTTTCAGTTTACGCAGGGTGCGGGCGATCACTTCCGGGTTGCTGGCAGCTTCGCCTTTCGGCGAAATGCTGTGGTATTGCTCGGCGGTGTTGGCCAGGTTCAGGAACTGACTAAACGCACGCGCAACGGGCAGCAGCTCGTCGTTCGACAAATTTTGTAAGGTGGTGAGCAACTCCTGGCGGTTAGCATCATTGCCAGCGCGTGAAGATTTCGACAACTTACGGATAGTTTCTACGCGTTCAAGAATGTGTTCTCCCAACGCATCCTTGATGGTTTCTCCCAGCACTTTGCCGAGCATACTGACATTACTACGCAATGCGGAATATTGTTCGTTCAT\n",
      ">b1101 NC_000913.3:1157868-1159302\n",
      "ATGTTTAAGAATGCATTTGCTAACCTGCAAAAGGTCGGTAAATCGCTGATGCTGCCGGTATCCGTACTGCCTATCGCAGGTATTCTGCTGGGCGTCGGTTCCGCGAATTTCAGCTGGCTGCCCGCCGTTGTATCGCATGTTATGGCAGAAGCAGGCGGTTCCGTCTTTGCAAACATGCCACTGATTTTTGCGATCGGTGTCGCCCTCGGCTTTACCAATAACGATGGCGTATCCGCGCTGGCCGCAGTTGTTGCCTATGGCATCATGGTTAAAACCATGGCCGTGGTTGCGCCACTGGTACTGCATTTACCTGCTGAAGAAATCGCCTCTAAACACCTGGCGGATACTGGCGTACTCGGAGGGATTATCTCCGGTGCGATCGCAGCGTACATGTTTAACCGTTTCTACCGTATTAAGCTGCCTGAGTATCTTGGCTTCTTTGCCGGTAAACGCTTTGTGCCGATCATTTCTGGCCTGGCTGCCATCTTTACTGGCGTTGTGCTGTCCTTCATTTGGCCGCCGATTGGTTCTGCAATCCAGACCTTCTCTCAGTGGGCTGCTTACCAGAACCCGGTAGTTGCGTTTGGCATTTACGGTTTCATCGAACGTTGCCTGGTACCGTTTGGTCTGCACCACATCTGGAACGTACCTTTCCAGATGCAGATTGGTGAATACACCAACGCAGCAGGTCAGGTTTTCCACGGCGACATTCCGCGTTATATGGCGGGTGACCCGACTGCGGGTAAACTGTCTGGTGGCTTCCTGTTCAAAATGTACGGTCTGCCAGCTGCCGCAATTGCTATCTGGCACTCTGCTAAACCAGAAAACCGCGCGAAAGTGGGCGGTATTATGATCTCCGCGGCGCTGACCTCGTTCCTGACCGGTATCACCGAGCCGATCGAGTTCTCCTTCATGTTCGTTGCGCCGATCCTGTACATCATCCACGCGATTCTGGCAGGCCTGGCATTCCCAATCTGTATTCTTCTGGGGATGCGTGACGGTACGTCGTTCTCGCACGGTCTGATCGACTTCATCGTTCTGTCTGGTAACAGCAGCAAACTGTGGCTGTTCCCGATCGTCGGTATCGGTTATGCGATTGTTTACTACACCATCTTCCGCGTGCTGATTAAAGCACTGGATCTGAAAACGCCGGGTCGTGAAGACGCGACTGAAGATGCAAAAGCGACAGGTACCAGCGAAATGGCACCGGCTCTGGTTGCTGCATTTGGTGGTAAAGAAAACATTACTAACCTCGACGCATGTATTACCCGTCTGCGCGTCAGCGTTGCTGATGTGTCTAAAGTGGATCAGGCCGGCCTGAAGAAACTGGGCGCAGCGGGCGTAGTGGTTGCTGGTTCTGGTGTTCAGGCGATTTTCGGTACTAAATCCGATAACCTGAAAACCGAGATGGATGAGTACATCCGTAACCACTAA\n",
      ">b2297 NC_000913.3:2414746-2416891\n",
      "GTGTCCCGTATTATTATGCTGATCCCTACCGGAACCAGCGTCGGTCTGACCAGCGTCAGCCTTGGCGTGATCCGTGCAATGGAACGCAAAGGCGTTCGTCTGAGCGTTTTCAAACCTATCGCTCAGCCGCGTACCGGTGGCGATGCGCCCGATCAGACTACGACTATCGTGCGTGCGAACTCTTCCACCACGACGGCCGCTGAACCGCTGAAAATGAGCTACGTTGAAGGTCTGCTTTCCAGCAATCAGAAAGATGTGCTGATGGAAGAGATCGTCGCAAACTACCACGCTAACACCAAAGACGCTGAAGTCGTTCTGGTTGAAGGTCTGGTCCCGACACGTAAGCACCAGTTTGCCCAGTCTCTGAACTACGAAATCGCTAAAACGCTGAATGCGGAAATCGTCTTCGTTATGTCTCAGGGCACTGACACCCCGGAACAGCTGAAAGAGCGTATCGAACTGACCCGCAACAGCTTCGGCGGTGCCAAAAACACCAACATCACCGGCGTTATCGTTAACAAACTGAACGCACCGGTTGATGAACAGGGTCGTACTCGCCCGGATCTGTCCGAGATTTTCGACGACTCTTCCAAAGCTAAAGTAAACAATGTTGATCCGGCGAAGCTGCAAGAATCCAGCCCGCTGCCGGTTCTCGGCGCTGTGCCGTGGAGCTTTGACCTGATCGCGACTCGTGCGATCGATATGGCTCGCCACCTGAATGCGACCATCATCAACGAAGGCGACATCAATACTCGCCGCGTTAAATCCGTCACTTTCTGCGCACGCAGCATTCCGCACATGCTGGAGCACTTCCGTGCCGGTTCTCTGCTGGTGACTTCCGCAGACCGTCCTGACGTGCTGGTGGCCGCTTGCCTGGCAGCCATGAACGGCGTAGAAATCGGTGCCCTGCTGCTGACTGGCGGTTACGAAATGGACGCGCGCATTTCTAAACTGTGCGAACGTGCTTTCGCTACCGGCCTGCCGGTATTTATGGTGAACACCAACACCTGGCAGACCTCTCTGAGCCTGCAGAGCTTCAACCTGGAAGTTCCGGTTGACGATCACGAACGTATCGAGAAAGTTCAGGAATACGTTGCTAACTACATCAACGCTGACTGGATCGAATCTCTGACTGCCACTTCTGAGCGCAGCCGTCGTCTGTCTCCGCCTGCGTTCCGTTATCAGCTGACTGAACTTGCGCGCAAAGCGGGCAAACGTATCGTACTGCCGGAAGGTGACGAACCGCGTACCGTTAAAGCAGCCGCTATCTGTGCTGAACGTGGTATCGCAACTTGCGTACTGCTGGGTAATCCGGCAGAGATCAACCGTGTTGCAGCGTCTCAGGGTGTAGAACTGGGTGCAGGGATTGAAATCGTTGATCCAGAAGTGGTTCGCGAAAGCTATGTTGGTCGTCTGGTCGAACTGCGTAAGAACAAAGGCATGACCGAAACCGTTGCCCGCGAACAGCTGGAAGACAACGTGGTGCTCGGTACGCTGATGCTGGAACAGGATGAAGTTGATGGTCTGGTTTCCGGTGCTGTTCACACTACCGCAAACACCATCCGTCCGCCGCTGCAGCTGATCAAAACTGCACCGGGCAGCTCCCTGGTATCTTCCGTGTTCTTCATGCTGCTGCCGGAACAGGTTTACGTTTACGGTGACTGTGCGATCAACCCGGATCCGACCGCTGAACAGCTGGCAGAAATCGCGATTCAGTCCGCTGATTCCGCTGCGGCCTTCGGTATCGAACCGCGCGTTGCTATGCTCTCCTACTCCACCGGTACTTCTGGTGCAGGTAGCGACGTAGAAAAAGTTCGCGAAGCAACTCGTCTGGCGCAGGAAAAACGTCCTGACCTGATGATCGACGGTCCGCTGCAGTACGACGCTGCGGTAATGGCTGACGTTGCGAAATCCAAAGCGCCGAACTCTCCGGTTGCAGGTCGCGCTACCGTGTTCATCTTCCCGGATCTGAACACCGGTAACACCACCTACAAAGCGGTACAGCGTTCTGCCGACCTGATCTCCATCGGGCCGATGCTGCAGGGTATGCGCAAGCCGGTTAACGACCTGTCCCGTGGCGCACTGGTTGACGATATCGTCTACACCATCGCGCTGACTGCGATTCAGTCTGCACAGCAGCAGTAA\n",
      ">b2296 NC_000913.3:2413469-2414672\n",
      "ATGTCGAGTAAGTTAGTACTGGTTCTGAACTGCGGTAGTTCTTCACTGAAATTTGCCATCATCGATGCAGTAAATGGTGAAGAGTACCTTTCTGGTTTAGCCGAATGTTTCCACCTGCCCGAAGCACGTATCAAATGGAAAATGGACGGCAATAAACAGGAAGCGGCTTTAGGTGCAGGCGCCGCTCACAGCGAAGCGCTCAACTTTATCGTTAATACTATTCTGGCACAAAAACCAGAACTGTCTGCGCAGCTGACTGCTATCGGTCACCGTATCGTACACGGCGGCGAAAAGTATACCAGCTCCGTAGTGATCGATGAGTCTGTTATTCAGGGTATCAAAGATGCAGCTTCTTTTGCACCGCTGCACAACCCGGCTCACCTGATCGGTATCGAAGAAGCTCTGAAATCTTTCCCACAGCTGAAAGACAAAAACGTTGCTGTATTTGACACCGCGTTCCACCAGACTATGCCGGAAGAGTCTTACCTCTACGCCCTGCCTTACAACCTGTACAAAGAGCACGGCATCCGTCGTTACGGCGCGCACGGCACCAGCCACTTCTATGTAACCCAGGAAGCGGCAAAAATGCTGAACAAACCGGTAGAAGAACTGAACATCATCACCTGCCACCTGGGCAACGGTGGTTCCGTTTCTGCTATCCGCAACGGTAAATGCGTTGACACCTCTATGGGCCTGACCCCGCTGGAAGGTCTGGTCATGGGTACCCGTTCTGGTGATATCGATCCGGCGATCATCTTCCACCTGCACGACACCCTGGGCATGAGCGTTGACGCAATCAACAAACTGCTGACCAAAGAGTCTGGCCTGCTGGGTCTGACCGAAGTGACCAGCGACTGCCGCTATGTTGAAGACAACTACGCGACGAAAGAAGACGCGAAGCGCGCAATGGACGTTTACTGCCACCGCCTGGCGAAATACATCGGTGCCTACACTGCGCTGATGGATGGTCGTCTGGACGCTGTTGTATTCACTGGTGGTATCGGTGAAAATGCCGCAATGGTTCGTGAACTGTCTCTGGGCAAACTGGGCGTGCTGGGCTTTGAAGTTGATCATGAACGCAACCTGGCTGCACGTTTCGGCAAATCTGGTTTCATCAACAAAGAAGGTACCCGTCCTGCGGTGGTTATCCCAACCAACGAAGAACTGGTTATCGCGCAAGACGCGAGCCGCCTGACTGCCTGA\n"
     ]
    }
   ],
   "source": [
    "li = []\n",
    "for i in id:\n",
    "    records = SeqIO.parse(gb_file, \"genbank\")\n",
    "    for record in records:\n",
    "        \n",
    "        gene_id,gene_seq,before_gene_seq = work(record, i) \n",
    "        temp = {'Id':f'{gene_id}_del',\n",
    "                   'Sequence upstream of the manipulation site (>100bp)':str(before_gene_seq),\n",
    "                   \"Reference sequence\":str(gene_seq),\n",
    "                   \"Inserted sequence\":\"-\",\n",
    "                   \"Manipulation type\":\"deletion\"}\n",
    "        li.append(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9719ec0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(li)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d59a21b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('4-18_input.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "302a5766",
   "metadata": {},
   "source": [
    "## 1.用户上传gb文件，可以将其生成fna基因组文件,支持真核与原核\n",
    "## 测试原核大肠、真核酵母的基因组gb文件生成基因组fna文件\n",
    "大肠基因组地址：ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
    "酵母基因组地址：ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.gbff.gz\n",
    "\n",
    "后端变更：\n",
    "    1.在数据预处理部分，生成基因组fna文件上传至上s3\n",
    "    2.将其用户的输入转换成标准的输入 \n",
    "\n",
    "\n",
    "## 2.同源臂长度优化问题：\n",
    "   1.增加针对序列合成的应用场景：当最大同源臂长度=最小同源臂长度时，同源臂的序列长度就是固定值。\n",
    "   2.原有使用引物批同源臂的场景：最大同源臂长度!=最小同源臂长度,会在此区间设计最佳引物，用于批同源臂，因此同源臂的长度是活动值。\n",
    "\n",
    "\n",
    "## 3.针对用户便捷性，增加用户提供编辑目标的坐标来执行此工具\n",
    "#### 1.编辑类型为敲除：通过Gene（基因id），Up region(默认值为0),Down region(默认值为0)，可实现任意位置目标片段的敲除\n",
    "#### 3.编辑类型为替换：通过Gene（基因id），Up region(默认值为0),Down region(默认值为0)，可实现任意位置目标片段的替换\n",
    "#### 2.编辑类型为插入：通过Gene（基因id），Up region(默认值为0),Down region(默认值为0)，可实现任意位置的插入\n",
    "用户上传一个基因组gb文件，和一个编辑信息文件csv文件（Name,Gene id,Up region(默认值为0),Down region(默认值为0)）------only_sgRNA\n",
    "用户上传一个基因组gb文件，和一个编辑信息文件csv文件（Name,Gene id,Up region(默认值为0),Down region(默认值为0),Inserted sequence,Manipulation type）------both_sgRNA_primer\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    ## 4.新增加功能：质粒合成订单的生成。\n",
    "\n",
    "\n",
    "    ## 5.基因组测序引物，由于之前直接由用户输入的同源臂上游默认是100，而在测序的时候为了保证测序的准确性需要同源臂上游序列设置的是200，所以出现问题。现在已经设置更改为200。\n",
    "\n",
    "\n",
    "    ## 6.基因组测序引物原来设计策略是：若目标序列小于600bp，只有一条引物。今天和师姐对接时，确定基因组测序引物要一对。\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "d8cb70e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9dde5fca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGTTTAAGAATGCATTTGCTAACCTGCAAAAGGTCGGTAAATCGCTGATGCTGCCGGTATCCGTACTGCCTATCGCAGGTATTCTGCTGGGCGTCGGTTCCGCGAATTTCAGCTGGCTGCCCGCCGTTGTATCGCATGTTATGGCAGAAGCAGGCGGTTCCGTCTTTGCAAACATGCCACTGATTTTTGCGATCGGTGTCGCCCTCGGCTTTACCAATAACGATGGCGTATCCGCGCTGGCCGCAGTTGTTGCCTATGGCATCATGGTTAAAACCATGGCCGTGGTTGCGCCACTGGTACTGCATTTACCTGCTGAAGAAATCGCCTCTAAACACCTGGCGGATACTGGCGTACTCGGAGGGATTATCTCCGGTGCGATCGCAGCGTACATGTTTAACCGTTTCTACCGTATTAAGCTGCCTGAGTATCTTGGCTTCTTTGCCGGTAAACGCTTTGTGCCGATCATTTCTGGCCTGGCTGCCATCTTTACTGGCGTTGTGCTGTCCTTCATTTGGCCGCCGATTGGTTCTGCAATCCAGACCTTCTCTCAGTGGGCTGCTTACCAGAACCCGGTAGTTGCGTTTGGCATTTACGGTTTCATCGAACGTTGCCTGGTACCGTTTGGTCTGCACCACATCTGGAACGTACCTTTCCAGATGCAGATTGGTGAATACACCAACGCAGCAGGTCAGGTTTTCCACGGCGACATTCCGCGTTATATGGCGGGTGACCCGACTGCGGGTAAACTGTCTGGTGGCTTCCTGTTCAAAATGTACGGTCTGCCAGCTGCCGCAATTGCTATCTGGCACTCTGCTAAACCAGAAAACCGCGCGAAAGTGGGCGGTATTATGATCTCCGCGGCGCTGACCTCGTTCCTGACCGGTATCACCGAGCCGATCGAGTTCTCCTTCATGTTCGTTGCGCCGATCCTGTACATCATCCACGCGATTCTGGCAGGCCTGGCATTCCCAATCTGTATTCTTCTGGGGATGCGTGACGGTACGTCGTTCTCGCACGGTCTGATCGACTTCATCGTTCTGTCTGGTAACAGCAGCAAACTGTGGCTGTTCCCGATCGTCGGTATCGGTTATGCGATTGTTTACTACACCATCTTCCGCGTGCTGATTAAAGCACTGGATCTGAAAACGCCGGGTCGTGAAGACGCGACTGAAGATGCAAAAGCGACAGGTACCAGCGAAATGGCACCGGCTCTGGTTGCTGCATTTGGTGGTAAAGAAAACATTACTAACCTCGACGCATGTATTACCCGTCTGCGCGTCAGCGTTGCTGATGTGTCTAAAGTGGATCAGGCCGGCCTGAAGAAACTGGGCGCAGCGGGCGTAGTGGTTGCTGGTTCTGGTGTTCAGGCGATTTTCGGTACTAAATCCGATAACCTGAAAACCGAGATGGATGAGTACATCCGTAACCACTAA\n",
      "1157868 1159302\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "('b1101',\n",
       " 'ATGTTTAAGAATGCATTTGCTAACCTGCAAAAGGTCGGTAAATCGCTGATGCTGCCGGTATCCGTACTGCCTATCGCAGGTATTCTGCTGGGCGTCGGTTCCGCGAATTTCAGCTGGCTGCCCGCCGTTGTATCGCATGTTATGGCAGAAGCAGGCGGTTCCGTCTTTGCAAACATGCCACTGATTTTTGCGATCGGTGTCGCCCTCGGCTTTACCAATAACGATGGCGTATCCGCGCTGGCCGCAGTTGTTGCCTATGGCATCATGGTTAAAACCATGGCCGTGGTTGCGCCACTGGTACTGCATTTACCTGCTGAAGAAATCGCCTCTAAACACCTGGCGGATACTGGCGTACTCGGAGGGATTATCTCCGGTGCGATCGCAGCGTACATGTTTAACCGTTTCTACCGTATTAAGCTGCCTGAGTATCTTGGCTTCTTTGCCGGTAAACGCTTTGTGCCGATCATTTCTGGCCTGGCTGCCATCTTTACTGGCGTTGTGCTGTCCTTCATTTGGCCGCCGATTGGTTCTGCAATCCAGACCTTCTCTCAGTGGGCTGCTTACCAGAACCCGGTAGTTGCGTTTGGCATTTACGGTTTCATCGAACGTTGCCTGGTACCGTTTGGTCTGCACCACATCTGGAACGTACCTTTCCAGATGCAGATTGGTGAATACACCAACGCAGCAGGTCAGGTTTTCCACGGCGACATTCCGCGTTATATGGCGGGTGACCCGACTGCGGGTAAACTGTCTGGTGGCTTCCTGTTCAAAATGTACGGTCTGCCAGCTGCCGCAATTGCTATCTGGCACTCTGCTAAACCAGAAAACCGCGCGAAAGTGGGCGGTATTATGATCTCCGCGGCGCTGACCTCGTTCCTGACCGGTATCACCGAGCCGATCGAGTTCTCCTTCATGTTCGTTGCGCCGATCCTGTACATCATCCACGCGATTCTGGCAGGCCTGGCATTCCCAATCTGTATTCTTCTGGGGATGCGTGACGGTACGTCGTTCTCGCACGGTCTGATCGACTTCATCGTTCTGTCTGGTAACAGCAGCAAACTGTGGCTGTTCCCGATCGTCGGTATCGGTTATGCGATTGTTTACTACACCATCTTCCGCGTGCTGATTAAAGCACTGGATCTGAAAACGCCGGGTCGTGAAGACGCGACTGAAGATGCAAAAGCGACAGGTACCAGCGAAATGGCACCGGCTCTGGTTGCTGCATTTGGTGGTAAAGAAAACATTACTAACCTCGACGCATGTATTACCCGTCTGCGCGTCAGCGTTGCTGATGTGTCTAAAGTGGATCAGGCCGGCCTGAAGAAACTGGGCGCAGCGGGCGTAGTGGTTGCTGGTTCTGGTGTTCAGGCGATTTTCGGTACTAAATCCGATAACCTGAAAACCGAGATGGATGAGTACATCCGTAACCACTAA',\n",
       " 1157868,\n",
       " 'NC_000913.3',\n",
       " 'plus')"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_seq_by_geneid(gb_file, gene_id='b1101', up_region=0, down_region=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "906ee46c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gb_path = '/home/yanghe/program/edit_sequence_design/input/only_primer/pMB1-Red-sgRNA-sacB-2297.gb'\n",
    "gb = SeqIO.read(gb_path, \"genbank\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cd856cb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "472a401a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a0d2ee9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dad3174a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1fe487c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e407edda",
   "metadata": {},
   "outputs": [],
   "source": [
    "def revComp(seq):\n",
    "    complementSeq=seq.translate(str.maketrans('ACGTacgtRYMKrymkVBHDvbhd', 'TGCAtgcaYRKMyrkmBVDHbvdh'))\n",
    "    revcompSeq = complementSeq[::-1]\n",
    "    return revcompSeq\n",
    "\n",
    "\n",
    "def conf_read(filename): \n",
    "    config = configparser.ConfigParser()\n",
    "    config.read(filename)\n",
    "    res = dict(config._sections[\"point\"])\n",
    "    return res\n",
    "\n",
    "def blast_search(input_file_path,genome,workdir):\n",
    "    \"\"\"\n",
    "    Arguments:\n",
    "        input_file_path[str]:  input information\n",
    "        genome[str]:  edit reference\n",
    "        workdir[str]: dir path\n",
    "    Return[dict]\n",
    "        {\n",
    "            \"key1\":{\n",
    "                \"chrom\":\"NC_001133.9\",\n",
    "                \"start\":\"1000\",\n",
    "                \"unique_mapped\":1,  100% comparison\n",
    "                \"mapped\":1,  frequency\n",
    "                \"reverse_mapped\":1, frequency\n",
    "                \"description\":\"NC_001133.9:26529-26708;NC_001133.9:26664-26843;\",  \n",
    "            },\n",
    "        }\n",
    "\n",
    "    \"\"\"\n",
    "    blast_output_file_path=os.path.join(workdir,'blast_search.txt')\n",
    "    ref_lib=genome.split('/')[-1].split('.')[0]\n",
    "    input_fasta = os.path.join(workdir,'blast.fasta')\n",
    "    fasta_length_dict = {}\n",
    "    my_records = []\n",
    "    with open(input_file_path,'r') as ifile:\n",
    "        index = 0 \n",
    "        for line in ifile:\n",
    "            if index != 0 :\n",
    "                linelist = line.split(',')\n",
    "                seqid = linelist[0]\n",
    "                seq = linelist[1].strip()\n",
    "                rec = SeqRecord(Seq(seq),id=seqid)\n",
    "\n",
    "                fasta_length_dict[seqid] = len(seq)\n",
    "                my_records.append(rec)\n",
    "            index += 1\n",
    "    # input fasta\n",
    "    SeqIO.write(my_records, input_fasta, \"fasta\")\n",
    "    # run blast\n",
    "    os.system(\"makeblastdb -in \"+genome+\" -dbtype nucl -parse_seqids -out \"+ref_lib)\n",
    "    os.system(\"blastn -query \"+input_fasta+\" -db \"+ref_lib+\" -outfmt 6 -task blastn -out \"+blast_output_file_path+\" -evalue 1e-30 \")\n",
    "\n",
    "    # return\n",
    "    dictall = {}\n",
    "    with open(blast_output_file_path,\"r\") as f:\n",
    "        for i in f:\n",
    "            linelist = i.split('\\t')\n",
    "            key = linelist[0]\n",
    "            chrom = linelist[1]\n",
    "            identity = linelist[2]\n",
    "            allength = linelist[3]\n",
    "            start = linelist[8]\n",
    "            end = linelist[9]\n",
    "            \n",
    "            if key not in dictall:\n",
    "                dictall[key] = {\n",
    "                    \"chrom\":chrom,\n",
    "                    \"start\":start,\n",
    "                    \"unique_mapped\":1 if (int(float(identity)) == 100 and int(float(allength))==fasta_length_dict[key]) else 0,\n",
    "                    \"mapped\":1,\n",
    "                    \"reverse_mapped\":1 if (int(start) > int(end) and int(float(identity)) == 100 and int(float(allength))==fasta_length_dict[key]) else 0,\n",
    "                    \"description\":'%s:%s-%s;' %(chrom,start,end),\n",
    "                }\n",
    "            else:\n",
    "                dictall[key][\"mapped\"] += 1\n",
    "                if int(float(identity)) == 100 and int(float(allength))==fasta_length_dict[key] :\n",
    "                    dictall[key][\"unique_mapped\"] += 1\n",
    "                    dictall[key][\"chrom\"] = chrom\n",
    "                    dictall[key][\"start\"] = start\n",
    "                    if int(start) > int(end):\n",
    "                        dictall[key][\"reverse_mapped\"] += 1\n",
    "                    dictall[key][\"description\"] += '%s:%s-%s;' %(chrom,start,end)\n",
    "    return dictall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "125e5549",
   "metadata": {},
   "outputs": [],
   "source": [
    "def input_to_primer_template(input_file_path, genome,workdir):\n",
    "    \"\"\"\n",
    "    Arguments:\n",
    "        input_file_path[str]:  input information\n",
    "        genome[str]:  edit reference\n",
    "        config[dic]: points information\n",
    "    Return [dict]\n",
    "        {\n",
    "            \"key1\":{\n",
    "                \"seq_uha_max_whole\":\"\",\n",
    "                \"seq_dha_max_whole\":\"\",\n",
    "                \"seq_altered\":\"\",\n",
    "                \"type\":\"\",   # [substitution,deletion,insertion]\n",
    "                \"ref\":\"\",\n",
    "                \"uha_upstream\": seq_uha_max_whole  up 100bp  sequence,\n",
    "                \"dha_downstream\":seq_dha_max_whole  down 100bp sequence,\n",
    "            },\n",
    "            \"key2\":{\n",
    "                \"seq_uha_max_whole\":\"\",\n",
    "                \"seq_dha_max_whole\":\"\",\n",
    "                \"seq_altered\":\"\",\n",
    "                \"type\":\"\",\n",
    "                \"ref\":\"\",\n",
    "                \"uha_upstream\": seq_uha_max_whole  up 100bp  sequence,\n",
    "                \"dha_downstream\":seq_dha_max_whole  down 100bp sequence,\n",
    "            }\n",
    "        }\n",
    "    \"\"\"\n",
    "    input_format=input_file_path.split('.')[-1]\n",
    "    with open(input_file_path,'r') as f:\n",
    "        input_header = f.readlines()[0]\n",
    "        # print(input_header)\n",
    "\n",
    "    primer_template = {}\n",
    "\n",
    "    # blast error file\n",
    "    with open(os.path.join(workdir,'error.txt'),'w') as blast_error_handler:\n",
    "        blast_error_handler.write('ID\\tERROR\\n')\n",
    "        if input_format != 'csv':\n",
    "            error_message = \"The input file format needs to be 'csv'\"\n",
    "            print(error_message)\n",
    "            return error_message\n",
    "        elif 'Sequence upstream' in input_header:\n",
    "            # input type 1: upstream\n",
    "            print('processing upstream input file ...')\n",
    "            record_dict = SeqIO.to_dict(SeqIO.parse(genome, \"fasta\"))\n",
    "            blast_search_dict = blast_search(input_file_path,genome,workdir)\n",
    "            df=pd.read_csv(input_file_path)\n",
    "\n",
    "            num_lines_df=df.shape[0]\n",
    "            for i in range(num_lines_df):\n",
    "                data=df.loc[i].values\n",
    "                #print(data)\n",
    "                mun_id=str(data[0])\n",
    "                if len(data) < 5:\n",
    "                    error_message = \"Some necessary input information is missing in the line \"+mun_id+\" of the input file\"\n",
    "                    print(error_message)\n",
    "                    return  error_message\n",
    "                else:\n",
    "                    upstream = data[1].strip().upper()\n",
    "                    ref = data[2]\n",
    "                    alt = data[3]\n",
    "                    mutation_type = data[4].strip().lower()\n",
    "                    name = mun_id\n",
    "                    \n",
    "                    if mun_id not in blast_search_dict:\n",
    "                        # no blast\n",
    "                        error_message = \"The upstream sequence of \" + mun_id + \" can not be mapped to the target genome. please check whether the target sequence is located on the target genome.\"\n",
    "                        blast_error_handler.write(mun_id + '\\t' +  error_message+ '\\n')\n",
    "                    elif blast_search_dict[mun_id][\"unique_mapped\"] > 1:\n",
    "                        # Compare 100 times\n",
    "                        error_message = \"The upstream sequence of \" + mun_id + \"  can be mapped to multiple loci in the target genome, %s, Please provide a longer upstream seqeunce.\" % blast_search_dict[mun_id][\"description\"]\n",
    "                        blast_error_handler.write(mun_id+'\\t'+ error_message+'\\n')\n",
    "                    elif blast_search_dict[mun_id][\"unique_mapped\"] == 0:\n",
    "                        # No 100 comparison\n",
    "\n",
    "                        error_message = \"The upstream sequence of \" + mun_id + \" can not be uniquely mapped to the target genome. Please check whether the target sequence is located on the target genome.\"\n",
    "                        blast_error_handler.write(mun_id+'\\t'+error_message+'\\n')\n",
    "                    elif blast_search_dict[mun_id][\"unique_mapped\"] == 1:\n",
    "                        # Index of the base starting to mutate on gene\n",
    "                        if blast_search_dict[mun_id][\"reverse_mapped\"]:\n",
    "                            record = revComp(str(record_dict[blast_search_dict[mun_id][\"chrom\"]].seq))\n",
    "                            upstream_start_index = len(record) - int(blast_search_dict[mun_id][\"start\"])\n",
    "                            strand = \"-\"  \n",
    "                        else:\n",
    "                            record = str(record_dict[blast_search_dict[mun_id][\"chrom\"]].seq)\n",
    "                            upstream_start_index = int(blast_search_dict[mun_id][\"start\"])-1\n",
    "                            strand = \"+\"\n",
    "                        chrom = blast_search_dict[mun_id][\"chrom\"]\n",
    "                        mutation_pos_index = upstream_start_index + len(upstream)\n",
    "                        \n",
    "                        res = create_mutation_info(\n",
    "                            record,mutation_type,mutation_pos_index,\n",
    "                            ref,alt,strand,chrom,\n",
    "                            name,mun_id\n",
    "                            )\n",
    "                        \n",
    "                        if isinstance(res,str):\n",
    "                            blast_error_handler.write(mun_id+'\\t'+res+'\\n')\n",
    "                        else:\n",
    "                            primer_template[mun_id] = res\n",
    "\n",
    "\n",
    "        elif 'Chr,Pos,Strand' in input_header:\n",
    "            # input type 2: vcf\n",
    "            print('processing vcf input file ...')\n",
    "            record_dict = SeqIO.to_dict(SeqIO.parse(genome, \"fasta\"))\n",
    "            df=pd.read_csv(input_file_path)\n",
    "            num_lines_df=df.shape[0]\n",
    "            for i in range(num_lines_df):\n",
    "                data=df.loc[i].values\n",
    "                #print(data)\n",
    "                mun_id=str(data[0])\n",
    "                if len(data) < 7:\n",
    "                    error_message = \"Some necessary input information is missing in the line \"+mun_id+\" of the input file\"\n",
    "                    print(error_message)\n",
    "                    return error_message\n",
    "                else:\n",
    "                    chrom = data[1].strip()\n",
    "                    pos = data[2]\n",
    "                    strand = data[3]\n",
    "                    ref = data[4].strip()\n",
    "                    alt = data[5]\n",
    "                    mutation_type = data[6]\n",
    "                    name = mun_id\n",
    "\n",
    "                    if strand == '+':\n",
    "                        record = str(record_dict[chrom].seq)\n",
    "                        mutation_pos_index = int(pos) - 1\n",
    "                    elif strand == '-':\n",
    "                        record = revComp(str(record_dict[chrom].seq))\n",
    "                        mutation_pos_index = len(record) - int(pos)\n",
    "                        \n",
    "                    # get mutation info dict\n",
    "                    res = create_mutation_info(\n",
    "                        record,mutation_type,mutation_pos_index,\n",
    "                        ref,alt,strand,chrom,\n",
    "                        name,mun_id\n",
    "                        )\n",
    "\n",
    "                    if isinstance(res,str):\n",
    "                        blast_error_handler.write(mun_id + '\\t'+ res +'\\n')\n",
    "                    else:\n",
    "                        primer_template[mun_id] = res\n",
    "        else:\n",
    "            error_message = \"The input file format not supported, Please rightly prepare input file for target manipulation as the example of 2,3-BD.\"\n",
    "            return  error_message\n",
    "    return primer_template\n",
    "\n",
    "\n",
    "def dict_to_df(dict_input_seq):\n",
    "    info_input_df = pd.DataFrame()\n",
    "    for item in dict_input_seq:\n",
    "        df = pd.DataFrame([dict_input_seq[item]])\n",
    "        df.insert(loc=0,value=item,column='id')\n",
    "        info_input_df = info_input_df.append(df)\n",
    "    info_input_df = info_input_df.reset_index(drop=True)\n",
    "    return info_input_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8b9d687a",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file_path='./input/editor_info.csv'\n",
    "genome_path='./input/GCA_000011325.1_ASM1132v1_genomic.fna'\n",
    "convert_input_file_chopchopInput_workdir =\"./output/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "51a7e017",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Sequence upstream of the manipulation site (&gt;100bp)</th>\n",
       "      <th>Reference sequence</th>\n",
       "      <th>Inserted sequence</th>\n",
       "      <th>Manipulation type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cgl0006_1176_G_A_sub</td>\n",
       "      <td>gtttccgctctgcgctgacctccctgatgaaccgctacgcacgtga...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>substitution</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cgl1436_1113_CAA_del</td>\n",
       "      <td>ACGACCACAAGAATTCCAAAACACCCGTTGACCAGGACAAATCCTC...</td>\n",
       "      <td>CAA</td>\n",
       "      <td>-</td>\n",
       "      <td>deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cgl2342_213_GCA_ins</td>\n",
       "      <td>agaacacatgagctccacgctcgcacgaaaagcttcaggcgacggt...</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA</td>\n",
       "      <td>insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cgl1790_1647_TCC_sub</td>\n",
       "      <td>GTGCTCAAATCCGCAACCATTTTCGCTACATCGACCTCGATGTAGT...</td>\n",
       "      <td>GAC</td>\n",
       "      <td>TTA</td>\n",
       "      <td>substitution</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cgl1386_327_18to15_sub</td>\n",
       "      <td>atggcgtcaatcgtctttgaaaacgtcacacgcaaatactctccgg...</td>\n",
       "      <td>gctgagatcgaaaagcgt</td>\n",
       "      <td>GCGTGGCGCGGTTAA</td>\n",
       "      <td>substitution</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cgl0591_-1_Ppgk_promoter_ins</td>\n",
       "      <td>ggcgatgatcggtaaaatcctcggcgcgcagatattgttcttgctg...</td>\n",
       "      <td>-</td>\n",
       "      <td>TTAATCAATACAATTGAATACCGGTGCCAGCGCCACACAATGTGTA...</td>\n",
       "      <td>insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cgl0141_cds_del</td>\n",
       "      <td>ATCACAACAGTCCCAATGACCTCAAAATCGGCCCCTTCTAGAATCG...</td>\n",
       "      <td>CTACATGCGATTCTTGGTGCGCGCAGTGGCAGGAGCGGTCCATGGA...</td>\n",
       "      <td>-</td>\n",
       "      <td>deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>153019_ecoil_ybeL_ins</td>\n",
       "      <td>ttgttaggagccaccacagacccgtagattcttgggcggtgtggtg...</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGAACAAGGTTGCTCAATATTACCGTGAACTGGTTGCGTCACTGA...</td>\n",
       "      <td>insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Cgl0851_ecoli_pgi_sub</td>\n",
       "      <td>caaacatcaacagttgggaagtagtgaacttcctcttcaacgacgg...</td>\n",
       "      <td>ctacctatttgcgcggtaccacttaatcagtgaatcagtggaagaa...</td>\n",
       "      <td>atgaaaaacatcaatccaacgcagaccgctgcctggcaggcactac...</td>\n",
       "      <td>substitution</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Id  \\\n",
       "0          Cgl0006_1176_G_A_sub   \n",
       "1          Cgl1436_1113_CAA_del   \n",
       "2           Cgl2342_213_GCA_ins   \n",
       "3          Cgl1790_1647_TCC_sub   \n",
       "4        Cgl1386_327_18to15_sub   \n",
       "5  Cgl0591_-1_Ppgk_promoter_ins   \n",
       "6               Cgl0141_cds_del   \n",
       "7         153019_ecoil_ybeL_ins   \n",
       "8         Cgl0851_ecoli_pgi_sub   \n",
       "\n",
       "  Sequence upstream of the manipulation site (>100bp)  \\\n",
       "0  gtttccgctctgcgctgacctccctgatgaaccgctacgcacgtga...    \n",
       "1  ACGACCACAAGAATTCCAAAACACCCGTTGACCAGGACAAATCCTC...    \n",
       "2  agaacacatgagctccacgctcgcacgaaaagcttcaggcgacggt...    \n",
       "3  GTGCTCAAATCCGCAACCATTTTCGCTACATCGACCTCGATGTAGT...    \n",
       "4  atggcgtcaatcgtctttgaaaacgtcacacgcaaatactctccgg...    \n",
       "5  ggcgatgatcggtaaaatcctcggcgcgcagatattgttcttgctg...    \n",
       "6  ATCACAACAGTCCCAATGACCTCAAAATCGGCCCCTTCTAGAATCG...    \n",
       "7  ttgttaggagccaccacagacccgtagattcttgggcggtgtggtg...    \n",
       "8  caaacatcaacagttgggaagtagtgaacttcctcttcaacgacgg...    \n",
       "\n",
       "                                  Reference sequence  \\\n",
       "0                                                  G   \n",
       "1                                                CAA   \n",
       "2                                                  -   \n",
       "3                                                GAC   \n",
       "4                                 gctgagatcgaaaagcgt   \n",
       "5                                                  -   \n",
       "6  CTACATGCGATTCTTGGTGCGCGCAGTGGCAGGAGCGGTCCATGGA...   \n",
       "7                                                  -   \n",
       "8  ctacctatttgcgcggtaccacttaatcagtgaatcagtggaagaa...   \n",
       "\n",
       "                                   Inserted sequence Manipulation type  \n",
       "0                                                  A      substitution  \n",
       "1                                                  -          deletion  \n",
       "2                                                GCA         insertion  \n",
       "3                                                TTA      substitution  \n",
       "4                                    GCGTGGCGCGGTTAA      substitution  \n",
       "5  TTAATCAATACAATTGAATACCGGTGCCAGCGCCACACAATGTGTA...         insertion  \n",
       "6                                                  -          deletion  \n",
       "7  ATGAACAAGGTTGCTCAATATTACCGTGAACTGGTTGCGTCACTGA...         insertion  \n",
       "8  atgaaaaacatcaatccaacgcagaccgctgcctggcaggcactac...      substitution  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(input_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4a493ec1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing upstream input file ...\n",
      "\n",
      "\n",
      "Building a new DB, current time: 03/03/2023 05:45:27\n",
      "New DB name:   /home/yanghe/program/data_preprocessing/GCA_000011325\n",
      "New DB title:  ./input/GCA_000011325.1_ASM1132v1_genomic.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/yanghe/program/data_preprocessing/GCA_000011325\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 1 sequences in 0.0530791 seconds.\n"
     ]
    }
   ],
   "source": [
    "dict_input_seq = input_to_primer_template(input_file_path,genome_path,convert_input_file_chopchopInput_workdir)\n",
    "info_input_df = dict_to_df(dict_input_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "191b3e8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seq_altered</th>\n",
       "      <th>type</th>\n",
       "      <th>ref</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_pos_index</th>\n",
       "      <th>geneid</th>\n",
       "      <th>name</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cgl0006_1176_G_A_sub</td>\n",
       "      <td>A</td>\n",
       "      <td>substitution</td>\n",
       "      <td>G</td>\n",
       "      <td>plus</td>\n",
       "      <td>6529</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0006_1176_G_A_sub</td>\n",
       "      <td>BA000036.3:6529-6530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cgl1436_1113_CAA_del</td>\n",
       "      <td></td>\n",
       "      <td>deletion</td>\n",
       "      <td>CAA</td>\n",
       "      <td>plus</td>\n",
       "      <td>1514089</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl1436_1113_CAA_del</td>\n",
       "      <td>BA000036.3:1514089-1514092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cgl2342_213_GCA_ins</td>\n",
       "      <td>GCA</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>plus</td>\n",
       "      <td>2484874</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl2342_213_GCA_ins</td>\n",
       "      <td>BA000036.3:2484874-2484875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cgl1790_1647_TCC_sub</td>\n",
       "      <td>TTA</td>\n",
       "      <td>substitution</td>\n",
       "      <td>GAC</td>\n",
       "      <td>plus</td>\n",
       "      <td>1899017</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl1790_1647_TCC_sub</td>\n",
       "      <td>BA000036.3:1899017-1899020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cgl1386_327_18to15_sub</td>\n",
       "      <td>GCGTGGCGCGGTTAA</td>\n",
       "      <td>substitution</td>\n",
       "      <td>gctgagatcgaaaagcgt</td>\n",
       "      <td>plus</td>\n",
       "      <td>1454441</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl1386_327_18to15_sub</td>\n",
       "      <td>BA000036.3:1454441-1454459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cgl0591_-1_Ppgk_promoter_ins</td>\n",
       "      <td>TTAATCAATACAATTGAATACCGGTGCCAGCGCCACACAATGTGTA...</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>plus</td>\n",
       "      <td>604469</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0591_-1_Ppgk_promoter_ins</td>\n",
       "      <td>BA000036.3:604469-604470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cgl0141_cds_del</td>\n",
       "      <td></td>\n",
       "      <td>deletion</td>\n",
       "      <td>CTACATGCGATTCTTGGTGCGCGCAGTGGCAGGAGCGGTCCATGGA...</td>\n",
       "      <td>plus</td>\n",
       "      <td>153222</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0141_cds_del</td>\n",
       "      <td>BA000036.3:153222-155613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>153019_ecoil_ybeL_ins</td>\n",
       "      <td>ATGAACAAGGTTGCTCAATATTACCGTGAACTGGTTGCGTCACTGA...</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>plus</td>\n",
       "      <td>153019</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>153019_ecoil_ybeL_ins</td>\n",
       "      <td>BA000036.3:153019-153020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Cgl0851_ecoli_pgi_sub</td>\n",
       "      <td>atgaaaaacatcaatccaacgcagaccgctgcctggcaggcactac...</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ctacctatttgcgcggtaccacttaatcagtgaatcagtggaagaa...</td>\n",
       "      <td>plus</td>\n",
       "      <td>907755</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0851_ecoli_pgi_sub</td>\n",
       "      <td>BA000036.3:907755-909378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             id  \\\n",
       "0          Cgl0006_1176_G_A_sub   \n",
       "1          Cgl1436_1113_CAA_del   \n",
       "2           Cgl2342_213_GCA_ins   \n",
       "3          Cgl1790_1647_TCC_sub   \n",
       "4        Cgl1386_327_18to15_sub   \n",
       "5  Cgl0591_-1_Ppgk_promoter_ins   \n",
       "6               Cgl0141_cds_del   \n",
       "7         153019_ecoil_ybeL_ins   \n",
       "8         Cgl0851_ecoli_pgi_sub   \n",
       "\n",
       "                                         seq_altered          type  \\\n",
       "0                                                  A  substitution   \n",
       "1                                                         deletion   \n",
       "2                                                GCA     insertion   \n",
       "3                                                TTA  substitution   \n",
       "4                                    GCGTGGCGCGGTTAA  substitution   \n",
       "5  TTAATCAATACAATTGAATACCGGTGCCAGCGCCACACAATGTGTA...     insertion   \n",
       "6                                                         deletion   \n",
       "7  ATGAACAAGGTTGCTCAATATTACCGTGAACTGGTTGCGTCACTGA...     insertion   \n",
       "8  atgaaaaacatcaatccaacgcagaccgctgcctggcaggcactac...  substitution   \n",
       "\n",
       "                                                 ref strand  \\\n",
       "0                                                  G   plus   \n",
       "1                                                CAA   plus   \n",
       "2                                                  -   plus   \n",
       "3                                                GAC   plus   \n",
       "4                                 gctgagatcgaaaagcgt   plus   \n",
       "5                                                  -   plus   \n",
       "6  CTACATGCGATTCTTGGTGCGCGCAGTGGCAGGAGCGGTCCATGGA...   plus   \n",
       "7                                                  -   plus   \n",
       "8  ctacctatttgcgcggtaccacttaatcagtgaatcagtggaagaa...   plus   \n",
       "\n",
       "   mutation_pos_index      geneid                          name  \\\n",
       "0                6529  BA000036.3          Cgl0006_1176_G_A_sub   \n",
       "1             1514089  BA000036.3          Cgl1436_1113_CAA_del   \n",
       "2             2484874  BA000036.3           Cgl2342_213_GCA_ins   \n",
       "3             1899017  BA000036.3          Cgl1790_1647_TCC_sub   \n",
       "4             1454441  BA000036.3        Cgl1386_327_18to15_sub   \n",
       "5              604469  BA000036.3  Cgl0591_-1_Ppgk_promoter_ins   \n",
       "6              153222  BA000036.3               Cgl0141_cds_del   \n",
       "7              153019  BA000036.3         153019_ecoil_ybeL_ins   \n",
       "8              907755  BA000036.3         Cgl0851_ecoli_pgi_sub   \n",
       "\n",
       "                       region  \n",
       "0        BA000036.3:6529-6530  \n",
       "1  BA000036.3:1514089-1514092  \n",
       "2  BA000036.3:2484874-2484875  \n",
       "3  BA000036.3:1899017-1899020  \n",
       "4  BA000036.3:1454441-1454459  \n",
       "5    BA000036.3:604469-604470  \n",
       "6    BA000036.3:153222-155613  \n",
       "7    BA000036.3:153019-153020  \n",
       "8    BA000036.3:907755-909378  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "415014a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8130498e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8bbc28d",
   "metadata": {},
   "outputs": [],
   "source": [
    "    \"uha_dha_config\": {\n",
    "        \"max_right_arm_seq_length\": 1050,\n",
    "        \"max_left_arm_seq_length\": 1050,\n",
    "        \"min_left_arm_seq_length\": 1000,\n",
    "        \"min_right_arm_seq_length\": 1000\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "afc560a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'id', 'seq_altered', 'type', 'ref', 'strand',\n",
       "       'mutation_pos_index', 'geneid', 'name', 'region', 'chrom_+_record'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_input_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "2be67436",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'id', 'seq_uha_max_whole', 'seq_dha_max_whole',\n",
       "       'seq_altered', 'type', 'ref', 'strand', 'mutation_pos_index', 'geneid',\n",
       "       'name', 'region', 'uha_upstream', 'dha_downstream'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b12d671",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_uha_max_whole\n",
    "seq_dha_max_whole\n",
    "mutation_pos_index\n",
    "uha_upstream\n",
    "dha_downstream\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "23730435",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.read_csv('/home/yanghe/program/edit_sequence_design/input/info_input.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c4938d0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "info_input_df = pd.read_csv('/home/yanghe/program/data_preprocessing/output/info_input.csv')\n",
    "info_input_df = del_Unnamed(info_input_df)\n",
    "genome_path='./input/GCA_000011325.1_ASM1132v1_genomic.fna'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "27a815b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_seq_from_genome(genome,gene_id,start=0,end=0):\n",
    "    # 读取基因组文件\n",
    "    records = SeqIO.parse(genome,'fasta')\n",
    "    \n",
    "    # 遍历基因组文件中的所有记录\n",
    "    for record in records:\n",
    "        # 如果当前记录的ID与所需的ID匹配\n",
    "        if record.id == gene_id:\n",
    "            # 提取坐标范围内的序列\n",
    "            if start ==0 and end ==0:\n",
    "                return str(record.seq)\n",
    "            else:\n",
    "                sequence = str(record.seq[start:end])\n",
    "                if sequence != '':\n",
    "                    return  sequence.upper()\n",
    "                else:\n",
    "                    return sequence\n",
    "\n",
    "def del_Unnamed(df):\n",
    "    \"\"\"\n",
    "    Deletes all the unnamed columns\n",
    "\n",
    "    :param df: pandas dataframe\n",
    "    \"\"\"\n",
    "    cols_del=[c for c in df.columns if 'Unnamed' in c]\n",
    "    return df.drop(cols_del,axis=1)\n",
    "\n",
    "\n",
    "def lambda2cols(df,lambdaf,in_coln,to_colns):         #apply函数的助手\n",
    "\n",
    "    if len(in_coln) == 2:\n",
    "        df_=df.apply(lambda x: lambdaf(x[in_coln[0]], x[in_coln[1]]),\n",
    "                     axis=1).apply(pd.Series)\n",
    "    if len(in_coln) == 3:\n",
    "        df_=df.apply(lambda x: lambdaf(x[in_coln[0]],x[in_coln[1]],x[in_coln[2]]),\n",
    "                     axis=1).apply(pd.Series)\n",
    "    elif len(in_coln) == 4:\n",
    "        df_=df.apply(lambda x: lambdaf(x[in_coln[0]],x[in_coln[1]],x[in_coln[2]],x[in_coln[3]]),\n",
    "                     axis=1).apply(pd.Series)\n",
    "    elif len(in_coln) == 5:\n",
    "         df_=df.apply(lambda x: lambdaf(x[in_coln[0]],x[in_coln[1]],x[in_coln[2]],x[in_coln[3]],x[in_coln[4]]),\n",
    "                     axis=1).apply(pd.Series)\n",
    "    df_.columns = to_colns\n",
    "    df = df.join(df_)        \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "94783bfc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ecc0fda",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_left_arm_seq_length=int(config['max_left_arm_seq_length'])\n",
    "max_right_arm_seq_length=int(config['max_right_arm_seq_length'])\n",
    "  # length \n",
    "\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "bc25c503",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seq_uha_max_whole</th>\n",
       "      <th>seq_dha_max_whole</th>\n",
       "      <th>seq_altered</th>\n",
       "      <th>type</th>\n",
       "      <th>ref</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_pos_index</th>\n",
       "      <th>geneid</th>\n",
       "      <th>name</th>\n",
       "      <th>region</th>\n",
       "      <th>uha_upstream</th>\n",
       "      <th>dha_downstream</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cgl0006_1176_G_A_sub</td>\n",
       "      <td>CTTGAAGGTCTTGAGGCGGTACGTAAGCGCCCGGGCATGTACATCG...</td>\n",
       "      <td>TTATTTCCGTGCGCGTTGGTGACCCACAGTTCGAAGGCCAGACCAA...</td>\n",
       "      <td>A</td>\n",
       "      <td>substitution</td>\n",
       "      <td>G</td>\n",
       "      <td>plus</td>\n",
       "      <td>6529</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0006_1176_G_A_sub</td>\n",
       "      <td>BA000036.3:6529-6530</td>\n",
       "      <td>GCGTGTAAGATGGAAACGTCTGTATCGGATAAGTAGCGAGGAGTGT...</td>\n",
       "      <td>AAGTTGGCCAATACGTGGGGTTGATTTTTTAAAACCAGACTGGCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cgl1436_1113_CAA_del</td>\n",
       "      <td>ATTAACGCCGCGCTCCCGCCGCGCCTTCAACAGCTCCTCCTGGGCC...</td>\n",
       "      <td>AGTGGTGGTGATGCGCTTCGGGATTCCCGCAGGTTCCGCCACTGCA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deletion</td>\n",
       "      <td>CAA</td>\n",
       "      <td>plus</td>\n",
       "      <td>1514089</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl1436_1113_CAA_del</td>\n",
       "      <td>BA000036.3:1514089-1514092</td>\n",
       "      <td>GCATGGCAGTCATGGCGATCAGCCTGCTGCTGTTTATCTAGCGTTC...</td>\n",
       "      <td>TCAATGGTGAGTTTCATGTTTTATCCAGCCTGACAAATAGAGAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cgl2342_213_GCA_ins</td>\n",
       "      <td>TCCAGAAGCTCAGGGACACCCATGCCAGTTTTACCGGACACGCGCA...</td>\n",
       "      <td>CCTGAAATTTCCGTCATGCCGACCGAATCAATGGCACGGCTGATCT...</td>\n",
       "      <td>GCA</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>plus</td>\n",
       "      <td>2484874</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl2342_213_GCA_ins</td>\n",
       "      <td>BA000036.3:2484874-2484875</td>\n",
       "      <td>GCCGCGGTAGGTGTCATAGACAGAGTCGAAAATCATCGCACGAGCT...</td>\n",
       "      <td>GACAATGAGGCACACAGCAGAGAAGGTAACAAAGGTCTGCACTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cgl1790_1647_TCC_sub</td>\n",
       "      <td>TCGATCAGCTTTCTCAACCTCATCGAGCAGGACAATAGCGTTACCC...</td>\n",
       "      <td>TTCTTCTTCGCGCCCAATAATCTCACGCACCGGTACAGCAACAGGC...</td>\n",
       "      <td>TTA</td>\n",
       "      <td>substitution</td>\n",
       "      <td>GAC</td>\n",
       "      <td>plus</td>\n",
       "      <td>1899017</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl1790_1647_TCC_sub</td>\n",
       "      <td>BA000036.3:1899017-1899020</td>\n",
       "      <td>TTGTCATGACGATATAGGTATTAAGAAAACTCACCTCACGGTTGTA...</td>\n",
       "      <td>TGCAGCAGCTTTCCCTGACTGCTCCGCTACAAATTCTTGAGCCTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cgl1386_327_18to15_sub</td>\n",
       "      <td>ACTACCTTTCTTGGACTTTCGAAGGCCCAATGTGGTCCCTCGGCGG...</td>\n",
       "      <td>GTTGCTGAAGCCTCACGCATTCTGCAGCTGGATCCGTATCTTGATC...</td>\n",
       "      <td>GCGTGGCGCGGTTAA</td>\n",
       "      <td>substitution</td>\n",
       "      <td>gctgagatcgaaaagcgt</td>\n",
       "      <td>plus</td>\n",
       "      <td>1454441</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl1386_327_18to15_sub</td>\n",
       "      <td>BA000036.3:1454441-1454459</td>\n",
       "      <td>GATCGTGGCCCAGAGTCATGGGAAGAGTTCTCCGAGTGGGGTCCAA...</td>\n",
       "      <td>GCCACGTTTAGTGCTTTCCACGAGGGCAAGATCGCGCCCACATTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cgl0591_-1_Ppgk_promoter_ins</td>\n",
       "      <td>GGAATCACAGGCAGGGTGGGAGATCTTCCACGCCTTGCAGGTGAAA...</td>\n",
       "      <td>ATGACTAAAACACTTGGTTCCCTTCAGCTGGAAGAAATCACGCTGA...</td>\n",
       "      <td>TTAATCAATACAATTGAATACCGGTGCCAGCGCCACACAATGTGTA...</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>plus</td>\n",
       "      <td>604469</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0591_-1_Ppgk_promoter_ins</td>\n",
       "      <td>BA000036.3:604469-604470</td>\n",
       "      <td>GGTTGGAGCGAAACTGCTCAAAGTAAATCACCGGGAAGATGCTCGA...</td>\n",
       "      <td>GCCAAGGGCGCTGCAGCAGTGTATGCCAATGACGTTTTCGTCCCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cgl0141_cds_del</td>\n",
       "      <td>GTGCGCTAATTGTATTCATCATCGGCGGCCACATGTGGTGGACCCT...</td>\n",
       "      <td>CAATTGCCTGTACATTAAGAATTTGCTGAGTGAACACATCTTTTTC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deletion</td>\n",
       "      <td>CTACATGCGATTCTTGGTGCGCGCAGTGGCAGGAGCGGTCCATGGA...</td>\n",
       "      <td>plus</td>\n",
       "      <td>153222</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0141_cds_del</td>\n",
       "      <td>BA000036.3:153222-155613</td>\n",
       "      <td>ACCGGCATGTTCCTCATCATGGCGTTCACCGCACTGCTCTCCCAAA...</td>\n",
       "      <td>AGTATTTCGGCAGTTCGTGCGGGGTCTGTAGGTCCAATTTGGTTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>153019_ecoil_ybeL_ins</td>\n",
       "      <td>CGGAACCGGCGAAAGCAAAGCCCTGCCCACCGGATGGAAACGCTGG...</td>\n",
       "      <td>GCTTTTAAGTTTTCTCGGTGATCCGTAACGAGAAGTTGAACAGCAA...</td>\n",
       "      <td>ATGAACAAGGTTGCTCAATATTACCGTGAACTGGTTGCGTCACTGA...</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>plus</td>\n",
       "      <td>153019</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>153019_ecoil_ybeL_ins</td>\n",
       "      <td>BA000036.3:153019-153020</td>\n",
       "      <td>GCGCCCTGGCTGCGAGTCTCATTGACAAACAAATCATGCGACCGCT...</td>\n",
       "      <td>AAAAGGTATTCTCCATTGTCTATTTTTCGGCCGATGAGCTGCGGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Cgl0851_ecoli_pgi_sub</td>\n",
       "      <td>ACACGCAATCGTATAACCCAAACCCTGCCCGAAGGAACTCAACGCA...</td>\n",
       "      <td>GAAAACTCCTTTATTGTCGTTAAATAACATCTTCAGGTTAGCTTGC...</td>\n",
       "      <td>atgaaaaacatcaatccaacgcagaccgctgcctggcaggcactac...</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ctacctatttgcgcggtaccacttaatcagtgaatcagtggaagaa...</td>\n",
       "      <td>plus</td>\n",
       "      <td>907755</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0851_ecoli_pgi_sub</td>\n",
       "      <td>BA000036.3:907755-909378</td>\n",
       "      <td>CGCCTCTAATAACGAAGAGTGTTGCAACCGCAAAAAGCACAAAGAT...</td>\n",
       "      <td>GGTGGTACAGCTTTTGTTGCAGGGCTTTGAGTTCGGGCAAGTCGGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             id  \\\n",
       "0          Cgl0006_1176_G_A_sub   \n",
       "1          Cgl1436_1113_CAA_del   \n",
       "2           Cgl2342_213_GCA_ins   \n",
       "3          Cgl1790_1647_TCC_sub   \n",
       "4        Cgl1386_327_18to15_sub   \n",
       "5  Cgl0591_-1_Ppgk_promoter_ins   \n",
       "6               Cgl0141_cds_del   \n",
       "7         153019_ecoil_ybeL_ins   \n",
       "8         Cgl0851_ecoli_pgi_sub   \n",
       "\n",
       "                                   seq_uha_max_whole  \\\n",
       "0  CTTGAAGGTCTTGAGGCGGTACGTAAGCGCCCGGGCATGTACATCG...   \n",
       "1  ATTAACGCCGCGCTCCCGCCGCGCCTTCAACAGCTCCTCCTGGGCC...   \n",
       "2  TCCAGAAGCTCAGGGACACCCATGCCAGTTTTACCGGACACGCGCA...   \n",
       "3  TCGATCAGCTTTCTCAACCTCATCGAGCAGGACAATAGCGTTACCC...   \n",
       "4  ACTACCTTTCTTGGACTTTCGAAGGCCCAATGTGGTCCCTCGGCGG...   \n",
       "5  GGAATCACAGGCAGGGTGGGAGATCTTCCACGCCTTGCAGGTGAAA...   \n",
       "6  GTGCGCTAATTGTATTCATCATCGGCGGCCACATGTGGTGGACCCT...   \n",
       "7  CGGAACCGGCGAAAGCAAAGCCCTGCCCACCGGATGGAAACGCTGG...   \n",
       "8  ACACGCAATCGTATAACCCAAACCCTGCCCGAAGGAACTCAACGCA...   \n",
       "\n",
       "                                   seq_dha_max_whole  \\\n",
       "0  TTATTTCCGTGCGCGTTGGTGACCCACAGTTCGAAGGCCAGACCAA...   \n",
       "1  AGTGGTGGTGATGCGCTTCGGGATTCCCGCAGGTTCCGCCACTGCA...   \n",
       "2  CCTGAAATTTCCGTCATGCCGACCGAATCAATGGCACGGCTGATCT...   \n",
       "3  TTCTTCTTCGCGCCCAATAATCTCACGCACCGGTACAGCAACAGGC...   \n",
       "4  GTTGCTGAAGCCTCACGCATTCTGCAGCTGGATCCGTATCTTGATC...   \n",
       "5  ATGACTAAAACACTTGGTTCCCTTCAGCTGGAAGAAATCACGCTGA...   \n",
       "6  CAATTGCCTGTACATTAAGAATTTGCTGAGTGAACACATCTTTTTC...   \n",
       "7  GCTTTTAAGTTTTCTCGGTGATCCGTAACGAGAAGTTGAACAGCAA...   \n",
       "8  GAAAACTCCTTTATTGTCGTTAAATAACATCTTCAGGTTAGCTTGC...   \n",
       "\n",
       "                                         seq_altered          type  \\\n",
       "0                                                  A  substitution   \n",
       "1                                                NaN      deletion   \n",
       "2                                                GCA     insertion   \n",
       "3                                                TTA  substitution   \n",
       "4                                    GCGTGGCGCGGTTAA  substitution   \n",
       "5  TTAATCAATACAATTGAATACCGGTGCCAGCGCCACACAATGTGTA...     insertion   \n",
       "6                                                NaN      deletion   \n",
       "7  ATGAACAAGGTTGCTCAATATTACCGTGAACTGGTTGCGTCACTGA...     insertion   \n",
       "8  atgaaaaacatcaatccaacgcagaccgctgcctggcaggcactac...  substitution   \n",
       "\n",
       "                                                 ref strand  \\\n",
       "0                                                  G   plus   \n",
       "1                                                CAA   plus   \n",
       "2                                                  -   plus   \n",
       "3                                                GAC   plus   \n",
       "4                                 gctgagatcgaaaagcgt   plus   \n",
       "5                                                  -   plus   \n",
       "6  CTACATGCGATTCTTGGTGCGCGCAGTGGCAGGAGCGGTCCATGGA...   plus   \n",
       "7                                                  -   plus   \n",
       "8  ctacctatttgcgcggtaccacttaatcagtgaatcagtggaagaa...   plus   \n",
       "\n",
       "   mutation_pos_index      geneid                          name  \\\n",
       "0                6529  BA000036.3          Cgl0006_1176_G_A_sub   \n",
       "1             1514089  BA000036.3          Cgl1436_1113_CAA_del   \n",
       "2             2484874  BA000036.3           Cgl2342_213_GCA_ins   \n",
       "3             1899017  BA000036.3          Cgl1790_1647_TCC_sub   \n",
       "4             1454441  BA000036.3        Cgl1386_327_18to15_sub   \n",
       "5              604469  BA000036.3  Cgl0591_-1_Ppgk_promoter_ins   \n",
       "6              153222  BA000036.3               Cgl0141_cds_del   \n",
       "7              153019  BA000036.3         153019_ecoil_ybeL_ins   \n",
       "8              907755  BA000036.3         Cgl0851_ecoli_pgi_sub   \n",
       "\n",
       "                       region  \\\n",
       "0        BA000036.3:6529-6530   \n",
       "1  BA000036.3:1514089-1514092   \n",
       "2  BA000036.3:2484874-2484875   \n",
       "3  BA000036.3:1899017-1899020   \n",
       "4  BA000036.3:1454441-1454459   \n",
       "5    BA000036.3:604469-604470   \n",
       "6    BA000036.3:153222-155613   \n",
       "7    BA000036.3:153019-153020   \n",
       "8    BA000036.3:907755-909378   \n",
       "\n",
       "                                        uha_upstream  \\\n",
       "0  GCGTGTAAGATGGAAACGTCTGTATCGGATAAGTAGCGAGGAGTGT...   \n",
       "1  GCATGGCAGTCATGGCGATCAGCCTGCTGCTGTTTATCTAGCGTTC...   \n",
       "2  GCCGCGGTAGGTGTCATAGACAGAGTCGAAAATCATCGCACGAGCT...   \n",
       "3  TTGTCATGACGATATAGGTATTAAGAAAACTCACCTCACGGTTGTA...   \n",
       "4  GATCGTGGCCCAGAGTCATGGGAAGAGTTCTCCGAGTGGGGTCCAA...   \n",
       "5  GGTTGGAGCGAAACTGCTCAAAGTAAATCACCGGGAAGATGCTCGA...   \n",
       "6  ACCGGCATGTTCCTCATCATGGCGTTCACCGCACTGCTCTCCCAAA...   \n",
       "7  GCGCCCTGGCTGCGAGTCTCATTGACAAACAAATCATGCGACCGCT...   \n",
       "8  CGCCTCTAATAACGAAGAGTGTTGCAACCGCAAAAAGCACAAAGAT...   \n",
       "\n",
       "                                      dha_downstream  \n",
       "0  AAGTTGGCCAATACGTGGGGTTGATTTTTTAAAACCAGACTGGCGT...  \n",
       "1  TCAATGGTGAGTTTCATGTTTTATCCAGCCTGACAAATAGAGAAAA...  \n",
       "2  GACAATGAGGCACACAGCAGAGAAGGTAACAAAGGTCTGCACTAAA...  \n",
       "3  TGCAGCAGCTTTCCCTGACTGCTCCGCTACAAATTCTTGAGCCTTA...  \n",
       "4  GCCACGTTTAGTGCTTTCCACGAGGGCAAGATCGCGCCCACATTTT...  \n",
       "5  GCCAAGGGCGCTGCAGCAGTGTATGCCAATGACGTTTTCGTCCCAG...  \n",
       "6  AGTATTTCGGCAGTTCGTGCGGGGTCTGTAGGTCCAATTTGGTTGT...  \n",
       "7  AAAAGGTATTCTCCATTGTCTATTTTTCGGCCGATGAGCTGCGGGA...  \n",
       "8  GGTGGTACAGCTTTTGTTGCAGGGCTTTGAGTTCGGGCAAGTCGGC...  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del_Unnamed(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d844ad32",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_left_arm_seq_length=int(config['max_left_arm_seq_length'])\n",
    "max_right_arm_seq_length=int(config['max_right_arm_seq_length'])\n",
    "\n",
    "def work(geneid, mutation_pos_index):\n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "6382d1ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seq_altered</th>\n",
       "      <th>type</th>\n",
       "      <th>ref</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_pos_index</th>\n",
       "      <th>geneid</th>\n",
       "      <th>name</th>\n",
       "      <th>region</th>\n",
       "      <th>chrom_+_record</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cgl0006_1176_G_A_sub</td>\n",
       "      <td>A</td>\n",
       "      <td>substitution</td>\n",
       "      <td>G</td>\n",
       "      <td>plus</td>\n",
       "      <td>6529</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0006_1176_G_A_sub</td>\n",
       "      <td>BA000036.3:6529-6530</td>\n",
       "      <td>GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cgl1436_1113_CAA_del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deletion</td>\n",
       "      <td>CAA</td>\n",
       "      <td>plus</td>\n",
       "      <td>1514089</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl1436_1113_CAA_del</td>\n",
       "      <td>BA000036.3:1514089-1514092</td>\n",
       "      <td>GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cgl2342_213_GCA_ins</td>\n",
       "      <td>GCA</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>plus</td>\n",
       "      <td>2484874</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl2342_213_GCA_ins</td>\n",
       "      <td>BA000036.3:2484874-2484875</td>\n",
       "      <td>GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cgl1790_1647_TCC_sub</td>\n",
       "      <td>TTA</td>\n",
       "      <td>substitution</td>\n",
       "      <td>GAC</td>\n",
       "      <td>plus</td>\n",
       "      <td>1899017</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl1790_1647_TCC_sub</td>\n",
       "      <td>BA000036.3:1899017-1899020</td>\n",
       "      <td>GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cgl1386_327_18to15_sub</td>\n",
       "      <td>GCGTGGCGCGGTTAA</td>\n",
       "      <td>substitution</td>\n",
       "      <td>gctgagatcgaaaagcgt</td>\n",
       "      <td>plus</td>\n",
       "      <td>1454441</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl1386_327_18to15_sub</td>\n",
       "      <td>BA000036.3:1454441-1454459</td>\n",
       "      <td>GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cgl0591_-1_Ppgk_promoter_ins</td>\n",
       "      <td>TTAATCAATACAATTGAATACCGGTGCCAGCGCCACACAATGTGTA...</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>plus</td>\n",
       "      <td>604469</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0591_-1_Ppgk_promoter_ins</td>\n",
       "      <td>BA000036.3:604469-604470</td>\n",
       "      <td>GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cgl0141_cds_del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deletion</td>\n",
       "      <td>CTACATGCGATTCTTGGTGCGCGCAGTGGCAGGAGCGGTCCATGGA...</td>\n",
       "      <td>plus</td>\n",
       "      <td>153222</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0141_cds_del</td>\n",
       "      <td>BA000036.3:153222-155613</td>\n",
       "      <td>GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>153019_ecoil_ybeL_ins</td>\n",
       "      <td>ATGAACAAGGTTGCTCAATATTACCGTGAACTGGTTGCGTCACTGA...</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>plus</td>\n",
       "      <td>153019</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>153019_ecoil_ybeL_ins</td>\n",
       "      <td>BA000036.3:153019-153020</td>\n",
       "      <td>GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Cgl0851_ecoli_pgi_sub</td>\n",
       "      <td>atgaaaaacatcaatccaacgcagaccgctgcctggcaggcactac...</td>\n",
       "      <td>substitution</td>\n",
       "      <td>ctacctatttgcgcggtaccacttaatcagtgaatcagtggaagaa...</td>\n",
       "      <td>plus</td>\n",
       "      <td>907755</td>\n",
       "      <td>BA000036.3</td>\n",
       "      <td>Cgl0851_ecoli_pgi_sub</td>\n",
       "      <td>BA000036.3:907755-909378</td>\n",
       "      <td>GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             id  \\\n",
       "0          Cgl0006_1176_G_A_sub   \n",
       "1          Cgl1436_1113_CAA_del   \n",
       "2           Cgl2342_213_GCA_ins   \n",
       "3          Cgl1790_1647_TCC_sub   \n",
       "4        Cgl1386_327_18to15_sub   \n",
       "5  Cgl0591_-1_Ppgk_promoter_ins   \n",
       "6               Cgl0141_cds_del   \n",
       "7         153019_ecoil_ybeL_ins   \n",
       "8         Cgl0851_ecoli_pgi_sub   \n",
       "\n",
       "                                         seq_altered          type  \\\n",
       "0                                                  A  substitution   \n",
       "1                                                NaN      deletion   \n",
       "2                                                GCA     insertion   \n",
       "3                                                TTA  substitution   \n",
       "4                                    GCGTGGCGCGGTTAA  substitution   \n",
       "5  TTAATCAATACAATTGAATACCGGTGCCAGCGCCACACAATGTGTA...     insertion   \n",
       "6                                                NaN      deletion   \n",
       "7  ATGAACAAGGTTGCTCAATATTACCGTGAACTGGTTGCGTCACTGA...     insertion   \n",
       "8  atgaaaaacatcaatccaacgcagaccgctgcctggcaggcactac...  substitution   \n",
       "\n",
       "                                                 ref strand  \\\n",
       "0                                                  G   plus   \n",
       "1                                                CAA   plus   \n",
       "2                                                  -   plus   \n",
       "3                                                GAC   plus   \n",
       "4                                 gctgagatcgaaaagcgt   plus   \n",
       "5                                                  -   plus   \n",
       "6  CTACATGCGATTCTTGGTGCGCGCAGTGGCAGGAGCGGTCCATGGA...   plus   \n",
       "7                                                  -   plus   \n",
       "8  ctacctatttgcgcggtaccacttaatcagtgaatcagtggaagaa...   plus   \n",
       "\n",
       "   mutation_pos_index      geneid                          name  \\\n",
       "0                6529  BA000036.3          Cgl0006_1176_G_A_sub   \n",
       "1             1514089  BA000036.3          Cgl1436_1113_CAA_del   \n",
       "2             2484874  BA000036.3           Cgl2342_213_GCA_ins   \n",
       "3             1899017  BA000036.3          Cgl1790_1647_TCC_sub   \n",
       "4             1454441  BA000036.3        Cgl1386_327_18to15_sub   \n",
       "5              604469  BA000036.3  Cgl0591_-1_Ppgk_promoter_ins   \n",
       "6              153222  BA000036.3               Cgl0141_cds_del   \n",
       "7              153019  BA000036.3         153019_ecoil_ybeL_ins   \n",
       "8              907755  BA000036.3         Cgl0851_ecoli_pgi_sub   \n",
       "\n",
       "                       region  \\\n",
       "0        BA000036.3:6529-6530   \n",
       "1  BA000036.3:1514089-1514092   \n",
       "2  BA000036.3:2484874-2484875   \n",
       "3  BA000036.3:1899017-1899020   \n",
       "4  BA000036.3:1454441-1454459   \n",
       "5    BA000036.3:604469-604470   \n",
       "6    BA000036.3:153222-155613   \n",
       "7    BA000036.3:153019-153020   \n",
       "8    BA000036.3:907755-909378   \n",
       "\n",
       "                                      chrom_+_record  \n",
       "0  GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...  \n",
       "1  GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...  \n",
       "2  GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...  \n",
       "3  GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...  \n",
       "4  GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...  \n",
       "5  GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...  \n",
       "6  GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...  \n",
       "7  GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...  \n",
       "8  GTGAGCCAGAACTCATCTTCTTTGCTCGAAACCTGGCGCCAAGTTG...  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d2cbc6f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6bbf94a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd14f22d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46e1fc17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09a1ceb6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "077e59a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56cde7c4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "774252d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98f43da7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21c89e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO \n",
    "import os,sys\n",
    "import warnings   \n",
    "warnings.filterwarnings('ignore')\n",
    "import configparser\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import argparse\n",
    "import json\n",
    "\n",
    "\n",
    "def revComp(seq):\n",
    "    complementSeq=seq.translate(str.maketrans('ACGTacgtRYMKrymkVBHDvbhd', 'TGCAtgcaYRKMyrkmBVDHbvdh'))\n",
    "    revcompSeq = complementSeq[::-1]\n",
    "    return revcompSeq\n",
    "\n",
    "\n",
    "def conf_read(filename): \n",
    "    config = configparser.ConfigParser()\n",
    "    config.read(filename)\n",
    "    res = dict(config._sections[\"point\"])\n",
    "    return res\n",
    "\n",
    "def input_to_primer_template(input_file_path, genome,config,workdir):\n",
    "    \"\"\"\n",
    "    Arguments:\n",
    "        input_file_path[str]:  input information\n",
    "        genome[str]:  edit reference\n",
    "        config[dic]: points information\n",
    "    Return [dict]\n",
    "        {\n",
    "            \"key1\":{\n",
    "                \"seq_uha_max_whole\":\"\",\n",
    "                \"seq_dha_max_whole\":\"\",\n",
    "                \"seq_altered\":\"\",\n",
    "                \"type\":\"\",   # [substitution,deletion,insertion]\n",
    "                \"ref\":\"\",\n",
    "                \"uha_upstream\": seq_uha_max_whole  up 100bp  sequence,\n",
    "                \"dha_downstream\":seq_dha_max_whole  down 100bp sequence,\n",
    "            },\n",
    "            \"key2\":{\n",
    "                \"seq_uha_max_whole\":\"\",\n",
    "                \"seq_dha_max_whole\":\"\",\n",
    "                \"seq_altered\":\"\",\n",
    "                \"type\":\"\",\n",
    "                \"ref\":\"\",\n",
    "                \"uha_upstream\": seq_uha_max_whole  up 100bp  sequence,\n",
    "                \"dha_downstream\":seq_dha_max_whole  down 100bp sequence,\n",
    "            }\n",
    "        }\n",
    "    \"\"\"\n",
    "    input_format=input_file_path.split('.')[-1]\n",
    "    with open(input_file_path,'r') as f:\n",
    "        input_header = f.readlines()[0]\n",
    "        # print(input_header)\n",
    "\n",
    "    primer_template = {}\n",
    "\n",
    "    # blast error file\n",
    "    with open(os.path.join(workdir,'error.txt'),'w') as blast_error_handler:\n",
    "        blast_error_handler.write('ID\\tERROR\\n')\n",
    "        if input_format != 'csv':\n",
    "            error_message = \"The input file format needs to be 'csv'\"\n",
    "            print(error_message)\n",
    "            return error_message\n",
    "        elif 'Sequence upstream' in input_header:\n",
    "            # input type 1: upstream\n",
    "            print('processing upstream input file ...')\n",
    "            record_dict = SeqIO.to_dict(SeqIO.parse(genome, \"fasta\"))\n",
    "            blast_search_dict = blast_search(input_file_path,genome,workdir)\n",
    "            df=pd.read_csv(input_file_path)\n",
    "\n",
    "            num_lines_df=df.shape[0]\n",
    "            for i in range(num_lines_df):\n",
    "                data=df.loc[i].values\n",
    "                #print(data)\n",
    "                mun_id=str(data[0])\n",
    "                if len(data) < 5:\n",
    "                    error_message = \"Some necessary input information is missing in the line \"+mun_id+\" of the input file\"\n",
    "                    print(error_message)\n",
    "                    return  error_message\n",
    "                else:\n",
    "                    upstream = data[1].strip().upper()\n",
    "                    ref = data[2]\n",
    "                    alt = data[3]\n",
    "                    mutation_type = data[4].strip().lower()\n",
    "                    name = mun_id\n",
    "                    \n",
    "                    if mun_id not in blast_search_dict:\n",
    "                        # no blast\n",
    "                        error_message = \"The upstream sequence of \" + mun_id + \" can not be mapped to the target genome. please check whether the target sequence is located on the target genome.\"\n",
    "                        blast_error_handler.write(mun_id + '\\t' +  error_message+ '\\n')\n",
    "                    elif blast_search_dict[mun_id][\"unique_mapped\"] > 1:\n",
    "                        # Compare 100 times\n",
    "                        error_message = \"The upstream sequence of \" + mun_id + \"  can be mapped to multiple loci in the target genome, %s, Please provide a longer upstream seqeunce.\" % blast_search_dict[mun_id][\"description\"]\n",
    "                        blast_error_handler.write(mun_id+'\\t'+ error_message+'\\n')\n",
    "                    elif blast_search_dict[mun_id][\"unique_mapped\"] == 0:\n",
    "                        # No 100 comparison\n",
    "\n",
    "                        error_message = \"The upstream sequence of \" + mun_id + \" can not be uniquely mapped to the target genome. Please check whether the target sequence is located on the target genome.\"\n",
    "                        blast_error_handler.write(mun_id+'\\t'+error_message+'\\n')\n",
    "                    elif blast_search_dict[mun_id][\"unique_mapped\"] == 1:\n",
    "                        # Index of the base starting to mutate on gene\n",
    "                        if blast_search_dict[mun_id][\"reverse_mapped\"]:\n",
    "                            record = revComp(str(record_dict[blast_search_dict[mun_id][\"chrom\"]].seq))\n",
    "                            upstream_start_index = len(record) - int(blast_search_dict[mun_id][\"start\"])\n",
    "                            strand = \"-\"  \n",
    "                        else:\n",
    "                            record = str(record_dict[blast_search_dict[mun_id][\"chrom\"]].seq)\n",
    "                            upstream_start_index = int(blast_search_dict[mun_id][\"start\"])-1\n",
    "                            strand = \"+\"\n",
    "                        chrom = blast_search_dict[mun_id][\"chrom\"]\n",
    "                        mutation_pos_index = upstream_start_index + len(upstream)\n",
    "\n",
    "\n",
    "                        # get mutation info dict\n",
    "                        res = create_mutation_info(\n",
    "                            record,mutation_type,mutation_pos_index,\n",
    "                            ref,alt,strand,chrom,\n",
    "                            name,mun_id,config\n",
    "                            )\n",
    "\n",
    "                            \n",
    "                        if isinstance(res,str):\n",
    "                            blast_error_handler.write(mun_id+'\\t'+res+'\\n')\n",
    "                        else:\n",
    "                            primer_template[mun_id] = res\n",
    "\n",
    "\n",
    "        elif 'Chr,Pos,Strand' in input_header:\n",
    "            # input type 2: vcf\n",
    "            print('processing vcf input file ...')\n",
    "            record_dict = SeqIO.to_dict(SeqIO.parse(genome, \"fasta\"))\n",
    "            df=pd.read_csv(input_file_path)\n",
    "            num_lines_df=df.shape[0]\n",
    "            for i in range(num_lines_df):\n",
    "                data=df.loc[i].values\n",
    "                #print(data)\n",
    "                mun_id=str(data[0])\n",
    "                if len(data) < 7:\n",
    "                    error_message = \"Some necessary input information is missing in the line \"+mun_id+\" of the input file\"\n",
    "                    print(error_message)\n",
    "                    return  error_message\n",
    "                else:\n",
    "                    chrom = data[1].strip()\n",
    "                    pos = data[2]\n",
    "                    strand = data[3]\n",
    "                    ref = data[4].strip()\n",
    "                    alt = data[5]\n",
    "                    mutation_type = data[6]\n",
    "                    name = mun_id\n",
    "\n",
    "                    if strand == '+':\n",
    "                        record = str(record_dict[chrom].seq)\n",
    "                        mutation_pos_index = int(pos) - 1\n",
    "                    elif strand == '-':\n",
    "                        record = revComp(str(record_dict[chrom].seq))\n",
    "                        mutation_pos_index = len(record) - int(pos)\n",
    "                    # get mutation info dict\n",
    "                    res = create_mutation_info(\n",
    "                        record,mutation_type,mutation_pos_index,\n",
    "                        ref,alt,strand,chrom,\n",
    "                        name,mun_id,config\n",
    "                        )\n",
    "                    if isinstance(res,str):\n",
    "                        blast_error_handler.write(mun_id + '\\t'+ res +'\\n')\n",
    "                    else:\n",
    "                        primer_template[mun_id] = res\n",
    "        else:\n",
    "            error_message = \"The input file format not supported, Please rightly prepare input file for target manipulation as the example of 2,3-BD.\"\n",
    "            return  error_message\n",
    "    return primer_template\n",
    "\n",
    "\n",
    "def blast_search(input_file_path,genome,workdir):\n",
    "    \"\"\"\n",
    "    Arguments:\n",
    "        input_file_path[str]:  input information\n",
    "        genome[str]:  edit reference\n",
    "        workdir[str]: dir path\n",
    "    Return[dict]\n",
    "        {\n",
    "            \"key1\":{\n",
    "                \"chrom\":\"NC_001133.9\",\n",
    "                \"start\":\"1000\",\n",
    "                \"unique_mapped\":1,  100% comparison\n",
    "                \"mapped\":1,  frequency\n",
    "                \"reverse_mapped\":1, frequency\n",
    "                \"description\":\"NC_001133.9:26529-26708;NC_001133.9:26664-26843;\",  \n",
    "            },\n",
    "        }\n",
    "\n",
    "    \"\"\"\n",
    "    blast_output_file_path=os.path.join(workdir,'blast_search.txt')\n",
    "    ref_lib=genome.split('/')[-1].split('.')[0]\n",
    "    input_fasta = os.path.join(workdir,'blast.fasta')\n",
    "    fasta_length_dict = {}\n",
    "    my_records = []\n",
    "    with open(input_file_path,'r') as ifile:\n",
    "        index = 0 \n",
    "        for line in ifile:\n",
    "            if index != 0 :\n",
    "                linelist = line.split(',')\n",
    "                seqid = linelist[0]\n",
    "                seq = linelist[1].strip()\n",
    "                rec = SeqRecord(Seq(seq),id=seqid)\n",
    "\n",
    "                fasta_length_dict[seqid] = len(seq)\n",
    "                my_records.append(rec)\n",
    "            index += 1\n",
    "    # input fasta\n",
    "    SeqIO.write(my_records, input_fasta, \"fasta\")\n",
    "    # run blast\n",
    "    os.system(\"makeblastdb -in \"+genome+\" -dbtype nucl -parse_seqids -out \"+ref_lib)\n",
    "    os.system(\"blastn -query \"+input_fasta+\" -db \"+ref_lib+\" -outfmt 6 -task blastn -out \"+blast_output_file_path+\" -evalue 1e-30 \")\n",
    "\n",
    "    # return\n",
    "    dictall = {}\n",
    "    with open(blast_output_file_path,\"r\") as f:\n",
    "        for i in f:\n",
    "            linelist = i.split('\\t')\n",
    "            key = linelist[0]\n",
    "            chrom = linelist[1]\n",
    "            identity = linelist[2]\n",
    "            allength = linelist[3]\n",
    "            start = linelist[8]\n",
    "            end = linelist[9]\n",
    "            \n",
    "            if key not in dictall:\n",
    "                dictall[key] = {\n",
    "                    \"chrom\":chrom,\n",
    "                    \"start\":start,\n",
    "                    \"unique_mapped\":1 if (int(float(identity)) == 100 and int(float(allength))==fasta_length_dict[key]) else 0,\n",
    "                    \"mapped\":1,\n",
    "                    \"reverse_mapped\":1 if (int(start) > int(end) and int(float(identity)) == 100 and int(float(allength))==fasta_length_dict[key]) else 0,\n",
    "                    \"description\":'%s:%s-%s;' %(chrom,start,end),\n",
    "                }\n",
    "            else:\n",
    "                dictall[key][\"mapped\"] += 1\n",
    "                if int(float(identity)) == 100 and int(float(allength))==fasta_length_dict[key] :\n",
    "                    dictall[key][\"unique_mapped\"] += 1\n",
    "                    dictall[key][\"chrom\"] = chrom\n",
    "                    dictall[key][\"start\"] = start\n",
    "                    if int(start) > int(end):\n",
    "                        dictall[key][\"reverse_mapped\"] += 1\n",
    "                    dictall[key][\"description\"] += '%s:%s-%s;' %(chrom,start,end)\n",
    "    return dictall\n",
    "\n",
    "\n",
    "def create_mutation_info(record,mutation_type,mutation_pos_index,ref,alt,strand,chrom,name,mun_id,config):\n",
    "    \"\"\"\n",
    "    Arguments:\n",
    "        record[str]: mutation site located genome\n",
    "        mutation_type[str]: deletion insertion substitution\n",
    "        mutation_pos_index[int]:  mutation_pos_index information\n",
    "        ref[str]:  mutation_pos_index ref\n",
    "        alt[str]: alter sequence\n",
    "        strand[str]:  + plus   - minus\n",
    "        chrom[str]:  mutation site located genome name\n",
    "        name[str]: mutation name\n",
    "        mun_id[str]: mutation id\n",
    "        config[dict]:  information\n",
    "    Return [dict]\n",
    "        {\n",
    "            \"seq_uha_max_whole\":\"\",\n",
    "            \"seq_dha_max_whole\":\"\",\n",
    "            \"seq_altered\":\"\",\n",
    "            \"type\":\"\",   # [substitution,deletion,insertion]\n",
    "            \"ref\":\"\",\n",
    "            \"uha_upstream\": seq_uha_max_whole  up 100bp  sequence,\n",
    "            \"dha_downstream\":seq_dha_max_whole  down 100bp sequence,\n",
    "        }\n",
    "    \n",
    "    \"\"\"\n",
    "    max_left_arm_seq_length=int(config['max_left_arm_seq_length'])\n",
    "    max_right_arm_seq_length=int(config['max_right_arm_seq_length'])\n",
    "\n",
    "    # length \n",
    "    if mutation_pos_index - max_left_arm_seq_length < 0:\n",
    "        error_message = \"The length of upstream sequence of manipulation site of \" + mun_id + \" must be larger than sum of 'Max Length of UHA' and 'Max Length of UIS'.\"\n",
    "        return error_message\n",
    "    \n",
    "    if mutation_type ==  \"insertion\":\n",
    "        info_dict = {\n",
    "            \"seq_uha_max_whole\":str(record[\n",
    "                mutation_pos_index - max_left_arm_seq_length : mutation_pos_index\n",
    "            ]),\n",
    "            \"seq_dha_max_whole\":str(record[\n",
    "                mutation_pos_index : mutation_pos_index + max_right_arm_seq_length\n",
    "            ]),\n",
    "            \"seq_altered\":alt,\n",
    "            \"type\":mutation_type,\n",
    "            \"ref\":ref,\n",
    "            \"strand\":\"plus\" if strand ==\"+\" else \"minus\",\n",
    "            \"mutation_pos_index\":mutation_pos_index,\n",
    "            \"geneid\":chrom,\n",
    "            \"name\":name,\n",
    "            \"region\":chrom+ ':' +  str(mutation_pos_index) +'-'+ str(int(mutation_pos_index)+len(ref)),\n",
    "            \"uha_upstream\":str(  \n",
    "                record[\n",
    "                    mutation_pos_index - max_left_arm_seq_length - 100 : mutation_pos_index - max_left_arm_seq_length\n",
    "                ]\n",
    "            ),\n",
    "            \"dha_downstream\":str(\n",
    "                record[\n",
    "                    mutation_pos_index + max_right_arm_seq_length  : mutation_pos_index + max_right_arm_seq_length  + 100\n",
    "                ]\n",
    "            ),\n",
    "\n",
    "        }\n",
    "        return info_dict      \n",
    "\n",
    "#         return info_dict\n",
    "    if mutation_type in [\"deletion\",\"substitution\"]:\n",
    "        genome_ref = record[mutation_pos_index:mutation_pos_index+len(ref)]\n",
    "        # print(genome_ref)\n",
    "        if genome_ref.upper() == ref.upper():\n",
    "            info_dict = {\n",
    "                \"seq_uha_max_whole\":str(record[\n",
    "                    mutation_pos_index \n",
    "                    - max_left_arm_seq_length :mutation_pos_index\n",
    "                    ]),\n",
    "                \"seq_dha_max_whole\":str(record[\n",
    "                    mutation_pos_index + len(ref)\n",
    "                    : mutation_pos_index + len(ref) + max_right_arm_seq_length\n",
    "                    ]),\n",
    "                \"seq_altered\":\"\" if alt=='-' else alt,\n",
    "                \"type\":mutation_type,\n",
    "                \"ref\":ref,\n",
    "                \"strand\":\"plus\" if strand == \"+\" else \"minus\",\n",
    "                \"mutation_pos_index\":mutation_pos_index,\n",
    "                \"geneid\":chrom,\n",
    "                \"name\":name,   \n",
    "                \"region\": chrom+ ':' +  str(mutation_pos_index) +'-'+ str(int(mutation_pos_index)+len(ref)),\n",
    "                \"uha_upstream\":str(  \n",
    "                    record[\n",
    "                        mutation_pos_index \n",
    "                    - max_left_arm_seq_length - 100\n",
    "                    : mutation_pos_index \n",
    "                    - max_left_arm_seq_length\n",
    "\n",
    "                    ]),\n",
    "                \"dha_downstream\":str(\n",
    "                    record[\n",
    "                        mutation_pos_index + len(ref) + max_right_arm_seq_length\n",
    "                        : mutation_pos_index + len(ref) + max_right_arm_seq_length + 100\n",
    "                    ]),\n",
    "            }\n",
    "            return info_dict\n",
    "        else:\n",
    "            error_message = \"The target mutation ref of \" + mun_id + \" can not be found in reference, please check.\"\n",
    "            return error_message\n",
    "    else:\n",
    "        error_message = \"The target manipulation type of \" + mun_id + \" must be equal to 'insertion,substitution or deletion', Please rightly prepare input file for target manipulation as the example of 2,3-BD.\"\n",
    "        return  error_message\n",
    "\n",
    "\n",
    "\n",
    "def dict_to_df(dict_input_seq):\n",
    "    info_input_df = pd.DataFrame()\n",
    "    for item in dict_input_seq:\n",
    "        df = pd.DataFrame([dict_input_seq[item]])\n",
    "        df.insert(loc=0,value=item,column='id')\n",
    "        info_input_df = info_input_df.append(df)\n",
    "    info_input_df = info_input_df.reset_index(drop=True)\n",
    "    return info_input_df\n",
    "\n",
    "\n",
    "def execute_input_2_chopchop_input(input_file_path, uha_dha_config, genome_path, convert_input_file_chopchopInput_workdir, chopchop_input):\n",
    "    crispr_info = pd.read_csv(input_file_path)\n",
    "    config = uha_dha_config\n",
    "    dict_input_seq = input_to_primer_template(input_file_path,genome_path,config, convert_input_file_chopchopInput_workdir)\n",
    "    info_input_df = dict_to_df(dict_input_seq)\n",
    "    info_input_df.to_csv(chopchop_input)\n",
    "\n",
    "\n",
    "\n",
    "def main(data):\n",
    "\n",
    "    input_file_path = data['input_file_path']\n",
    "    uha_dha_config = data['uha_dha_config']\n",
    "    genome_path = data['ref_genome']\n",
    "    convert_input_file_chopchopInput_workdir = data['data_preprocessing_workdir']\n",
    "    \n",
    "    if not os.path.exists(convert_input_file_chopchopInput_workdir):\n",
    "        os.makedirs(convert_input_file_chopchopInput_workdir)\n",
    "    chopchop_input =os.path.join(\n",
    "        data['data_preprocessing_workdir'],\n",
    "        'info_input.csv'\n",
    "    )\n",
    "    \n",
    "    execute_input_2_chopchop_input(input_file_path, uha_dha_config, genome_path, convert_input_file_chopchopInput_workdir, chopchop_input)\n",
    "    return chopchop_input\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('--input', '-i', help='input params file', required=True) \n",
    "    args = parser.parse_args()\n",
    "    input_path =  args.input\n",
    "    with open(input_path, \"r\") as f:\n",
    "        data = json.load(f)\n",
    "\n",
    "    main(data)  \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "vscode": {
   "interpreter": {
    "hash": "0ce5cabab122b3db5c05bbe8f8ff7fd3aa7bcecef663505ad3dd84494ebc9b75"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
